---
title: "Trustworthy and Adaptable Data Integration"
author: "Kris Sankaran"
output:
  xaringan::moon_reader:
    css: ["default", "css/xaringan-themer.css"]
    lib_dir: libs
    self_contained: false
    nature:
      highlightStyle: github
      highlightLines: true
      ratio: "16:9"
    seal: false
---

class: title

```{r, echo = FALSE, warning = FALSE, message = FALSE}
library(MASS)
library(knitr)
library(RefManageR)
library(tidyverse)
opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE, dpi = 200, fig.align = "center", fig.width = 6, fig.height = 3)
min_theme <- theme_minimal() + 
  theme(
    panel.grid.minor = element_blank(),
    panel.background = element_rect(fill = "#f7f7f7"),
    panel.border = element_rect(fill = NA, color = "#0c0c0c", size = 0.6),
    axis.text = element_text(size = 14),
    strip.text = element_text(size = 16),
    axis.title = element_text(size = 16),
    legend.position = "bottom"
  )
theme_set(min_theme)

# overwrite some default scales in ggplot2
scale_fill_continuous <- function(...) scico::scale_fill_scico(..., palette = "lapaz", direction = -1)
scale_colour_discrete <- function(...) ggplot2::scale_color_brewer(..., palette = "Set2")
scale_x_continuous <- function(...) ggplot2::scale_x_continuous(..., expand = c(0, 0))
scale_y_continuous <- function(...) ggplot2::scale_y_continuous(..., expand = c(0, 0))

BibOptions(
  check.entries = TRUE, 
  bib.style = "authoryear", 
  cite.style = "authoryear", 
  style = "markdown",
  hyperlink = FALSE, 
  dashed = FALSE,
  max.names = 1
)
bib <- ReadBib("references.bib")
```

<div id="title">
Trustworthy and Adaptable Data Integration
</div>
<div id="under_title">
</div>

<div id="subtitle_right">
IMS - NUS Workshop Proposal <br/>
03 | July | 2024 <br/>
Slides: <a href="https://go.wisc.edu/54tmr9">go.wisc.edu/54tmr9</a>
</div>

<div id="subtitle">
Kris Sankaran, Wei-Yin Loh <br/> 
Susan Holmes Wanjie Wang <br/>
Bibhas Chakraborty, Bee Choo Tai
</div>

---

### Outline

.pull-left[
**Setting**

1. Scientific Context
1. Statistical Challenges
1. Mathematical Opportunities
]

.pull-right[
**Structure**

1. Values
1. Audience
1. Program
]

---

class: middle

.center[
## Scientific Context
]

---

### Data Integration

For many biological problems, it is necessary to gather complementary views,

1. Horizontal: Gather the same type of measurements across related, but not
identical cohorts.
1. Vertical: Gather multiple types of measurements from the same biological systems.

Research in data integration should make it easier to answer questions that need
data from all these sources.

---

### Example: Cancer Research

Information about the molecular activity and global organization of cells has
helped advance understanding and personalized treatment of cancer.

---

### Example: Gene Editing

(?)

---

class: middle

.center[
## Statistical Challenges
]

---

### Adaptability

How can we enable careful data analysis when each problem might involve diverse
data types and experimental designs?

1. General-purpose techniques
1. Effective diagnostics
1. Strategies for scalability

Biological insights can be hindered by inappropriate assumptions or the failure
to use all of the data.

---

### Reproducibility

Complex workflows -- which have many interlocking steps and may use black boxes
-- challenge scientific reproducibility and interpretability.

1. Assessing stability
1. Addressing selection baises
1. Supporting re-usable software

---

### Community Resources

Larger community databases can be informative, but may be noisier or only
indirectly related to questions of interest. How should they be brought into a
specific analysis?

1. Cell type atlases
1. Gene function annotation
1. Electronic health records

---

class: middle

.center[
## Mathematical Opportunities
]

---

### Motivation

Theoretical analysis has driven advances in biological data integration,

.pull-left[
1. **Improved Algorithms: Theory can provide richer vocabulary for formulating and answering quantitative questions based on biological data.**

1. Consolidation: Theory can create coherent framework from which to evaluate
available methods.
]

.pull-right[
  High-Dimensional CCA
]

---

### Motivation

Theoretical analysis has driven advances in biological data integration,

.pull-left[
1. Improved Algorithms: Theory can provide richer vocabulary for formulating and answering quantitative questions based on biological data.

1. **Consolidation: Theory can create coherent framework from which to evaluate available methods.**
]

.pull-right[
  Tensor methods for network integration
]

---

### Motivation

Modern data can also inspire developments in theory,

.pull-left[
1. **Expanded Abstractions: As the collection of applications grows, theoretical concepts can be enriched.**

1. Revised Assumptions: Properties that are taken for granted in narrower analysis settings may no longer apply.
]

.pull-right[
  Empirical Bayes PCA
]
---

### Motivation

Modern data can also inspire developments in theory,

.pull-left[
1. Expanded Abstractions: As the collection of applications grows, theoretical concepts can be enriched.
1. **Revised Assumptions: Properties that are taken for granted in narrower analysis settings may no longer apply.**
]

.pull-right[
  Testing for the alignability of multiple tables
]

---

class: middle

.center[
## Workshop Structure
]

---

### Goals

1. Support discussions that could lead to methodological and conceptual
breakthroughs in biological data integration.

1. Provide training on current best practices for researchers who are entering
the field either from statistical or biological backgrounds.

---

### Values: Communication

Our workshop aims to promote effective communication between researchers working
on complementary biological and statistical aspects of the data integration
problem.

1. Focused discussions will clarify challenges and identify strategies for
interdisciplinary research.

1. Tutorials will help establish shared vocabulary that can serve as a reference
throughout the workshop.

---

### Values: Hands-on Learning

We want workshop participants with varying levels of expertise to leave with a
concrete understanding of the current theory and practice of data integration.

1. We will curate a collection of problems, datasets, and methods for data
integration which can serve as a resource for the wider community.

1. We will create opportunities for participants to present short research demos
and talks.

---

### Schedule

The program will span roughly three weeks, divided into two main components.

Draw a timeline with:

2 x 4 days: Tutorials

Facilitator training

8 days: Talks and Discussion 
Talks + Discussion

---

### Tutorials: Biological Data Management and Visualization

**Kris Sankaran** will provide an introduction to managing and visualizing
multi-omics data using tools from the Bioconductor ecosystem.

---

### Tutorials: Introduction to Regression Trees

**Wei-Yin Loh** will offer a course on tree-based methods and their application
*to mixed data types that arise in biomedical settings.

---

### Tutorials: Generative Models for Multi-omics

**Kris Sankaran** will review concepts from probabilistic modeling, evaluation,
and interpretation that are the basis for many data integration methods.

---

### Tutorials: Case Studies from the CPTAC Project

**Xiaoyu Song** will lead a session describing how problems have been formulated
*and solved in her expeience with the Clinical Proteomic Tumor Analsyis
*Consortium (CPTAC).

---

### Talk Themes

Talks will be grouped according to methodological themes.

.pull-left[
* Robust and Reproducible Integration

* Spatial and Single-Cell Analysis

* Efficient and Scalable Models
]

.pull-right[
* Emerging Sequencing Technologies

* AI for Biological Foundation Models
]

---

### Interaction

Beyond the featured talks, we will create opportunities for both formal and
informal interaction among all participants.

1. Facilitated discussion
1. Lightning research talks
1. Data analysis demos 

Depending on speaker preferences, we will curate resources from these activities
onto public/private links on a workshop website.

---

### Recruitment

The organizing team will recruit students, scientists, and researchers locally
and abroad who have the potential to benefit from one another's expertise.

1. University Campuses (50%)
1. Research Institutes (20%)
1. Hospitals (20%)
1. Industry (10%)

---

### Why IMS-NUS?

1. Interdisciplinary emphasis: The workshop relies on an environment that allows
researchers from multiple backgounds to feel welcome.

1. Statistics reputation: Unlike related workshops, we hope to emphasize the
role of statistical thinking in integration, so the IMS - NUS connection will
help.

1. Location: The strength of both the statistics and bioinformatics communities
in Singapore will ensure that the workshop balances theoretical and applied
perspectives.

---

### Conclusion

1. We believe that this topic has the potential both to demonstrate the value of
statistics to the wider scientific community and to bring exciting, new problems
into statistical discussion.

1. Our team is excited by the possibility of contributing to the IMS - NUS
workshop series.

---

### References

```{r, results='asis', echo = FALSE}
PrintBibliography(bib, start = 1, end = 4)
```

---

```{r, results='asis', echo = FALSE}
PrintBibliography(bib, start = 5, end = 8)
```
