<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Trustworthy and Adaptable Biological Data Integration</title>
    <meta charset="utf-8" />
    <meta name="author" content="Kris Sankaran" />
    <script src="libs/header-attrs-2.27/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">


class: title



&lt;div id="title"&gt;
Trustworthy and Adaptable Biological Data Integration&lt;br/&gt;
&lt;br/&gt;
&lt;/div&gt;
&lt;div id="under_title"&gt;
IMS - NUS Workshop Proposal 
&lt;/div&gt;

&lt;div id="subtitle_right"&gt;
03 | July | 2024 &lt;br/&gt;
Slides: &lt;a href="https://go.wisc.edu/8k8r2q"&gt;go.wisc.edu/8k8r2q&lt;/a&gt;
&lt;/div&gt;

&lt;div id="subtitle"&gt;
&lt;i&gt;Organizing Team&lt;/i&gt;&lt;br/&gt;
&lt;span style="font-size: 18px;"&gt;
Kris Sankaran, UWM&lt;br/&gt;
Wei-Yin Loh, UWM&lt;br/&gt; 
Susan Holmes, Stanford&lt;br/&gt;
Bibhas Chakraborty, Duke-NUS&lt;br/&gt;
Bee Choo Tai, NUS &lt;br/&gt;
Wanjie Wang, NUS &lt;br/&gt;
&lt;/span&gt;
&lt;/div&gt;

---

class: middle

.center[
## Scientific Context
]

---

### Data Integration

For many biological problems, researchers gather data from multiple views,

.center[
&lt;img src="figures/integration_types.png" width=1000/&gt;
]

Data integration makes it easier to answer questions using all relevant data.

---

### Example: Cancer

.pull-left[
1. Data about molecular activity and cellular organization has clarified
when treatments are more likely to be effective. 

1. Several technologies and levels of resolution -- population,
person, tissue, cell -- need to be studied together.
]

.pull-right[
&lt;img src="figures/Lung_cancer_HAVCR2.jpg" width=450/&gt;
&lt;span style="font-size: 22px;"&gt;
Cancer cells in liver tissue, from [3].
&lt;/span&gt;
]

---

### Example: COVID-19

1. Molecular data have shed light on the immune system mechanisms responsible
for variable COVID-19 severity.

1. Multiple high-dimensional sources can be integrated to identify key molecular
interactions.

.center[
&lt;img src="figures/covid_multiomics.jpg" width=550/&gt; &lt;br/&gt;
&lt;span style="font-size: 22px;"&gt;
Graphical summary from [10].
&lt;/span&gt;
]


---

class: middle

.center[
## Statistical Challenges
]

---

### Adaptability

How can we enable careful statistical analysis across diverse data types and
experimental designs?

.pull-left[
1. General-purpose techniques
1. Effective diagnostics
1. Scalability
]

.pull-right[
&lt;img src="figures/data_types.png" width=500/&gt;
]

Biological insights can be hindered by inappropriate assumptions or the failure
to use all of the data.

---

### Reproducibility

Complex workflows which have many interlocking steps and may use black boxes
challenge scientific reproducibility and interpretability. The community needs
methods for,

.pull-left[
1. Assessing stability
1. Addressing selection baises
1. Supporting re-usable software
]

.pull-right[
&lt;img src="figures/multimodal_flow.png" width=500/&gt;
]

---

### Community Resources

Larger community databases can be informative, but may be noisier or only
indirectly related to questions of interest. How should they be brought into a
specific analysis?

.pull-left[
1. Cell type atlases
1. Gene function annotation
1. DNA foundation models
]

.pull-right[
&lt;img src="figures/NeuronsGraphic-Lee_Dalley.png" width=700/&gt;
]

---

class: middle

.center[
## Mathematical Opportunities
]

---

### Motivation

Theoretical analysis has driven advances in biological data integration,

.pull-left[
1. **Improved Algorithms: Theory creates language for forming and answering novel questions.**

1. Consolidation: Theory can create coherent framework from which to view
available methods.
]

.pull-right[
  &lt;img src="figures/cnv.png" width=440/&gt;
  &lt;span style="font-size: 18px;"&gt;
  Scan statistics offer a useful lens for analyzing copy number variation [5].
  &lt;/span&gt;
]

---

### Motivation

Theoretical analysis has driven advances in biological data integration,

.pull-left[
1. Improved Algorithms: Theory creates language for forming and answering novel questions.

1. **Consolidation: Theory can create coherent framework from which to view available methods.**
]

.pull-right[
  &lt;img src="figures/cfit.jpg" width=480/&gt;
  &lt;span style="font-size: 18px;"&gt;
  Low-rank models are ubiquitous in integration across high-dimensional genomics
  data [4].
  &lt;/span&gt;
]


---

### Motivation

Modern data can also inspire developments in theory,

.pull-left[
1. **Expanded Abstractions: As more applications emerge, theoretical concepts can be enriched.**

1. Revised Assumptions: Properties that are taken for granted in narrower analysis settings may no longer apply.
]

.pull-right[
  .center[
&lt;img src="figures/lfdr.png" width=240/&gt;
]

&lt;span style="font-size: 18px;"&gt;
Microarray analysis sparked developments in empirical bayes and hypothesis testing [6].
&lt;/span&gt;
]
---

### Motivation

Modern data can also inspire developments in theory,

.pull-left[
1. Expanded Abstractions: As more applications emerge, theoretical concepts can be enriched.

1. **Revised Assumptions: Properties that are taken for granted in narrower analysis settings may no longer apply.**
]

.pull-right[
&lt;img src="figures/sparse_gca.png"/&gt;
&lt;span style="font-size: 18px;"&gt;
Problems from data integration can help inspire research in high-dimensional
statistics [7].
&lt;/span&gt;
]

---

class: middle

.center[
## Workshop Structure
]

---

### Goals

&lt;img src="figures/dna-molecule.png" width=40 align="left" hspace=20/&gt;Support discussions that could shape methodological progress in biological data integration.
&lt;br/&gt;
&lt;br/&gt;
&lt;br/&gt;

&lt;img src="figures/normal-distn.png" width=80 align="left" hspace=10/&gt; Introduce cutting-edge biological datasets and problems to researchers from statistics.
&lt;br/&gt;
&lt;br/&gt;
&lt;br/&gt;

&lt;img src="figures/computer.png" width=80 align="left" hspace=10/&gt;Provide training on current best practices for researchers who are entering the field either from statistical or biological backgrounds.

---

### Values: Interdisciplinarity

Our workshop aims to promote effective communication between researchers working
on complementary biological and statistical aspects of the data integration
problem.

.pull-three-quarters-left[
1. Focused discussions will clarify challenges and identify strategies for
interdisciplinary research.

1. Tutorials will help establish shared vocabulary that can serve as a reference
throughout the workshop.
]
.pull-three-quarters-right[
&lt;img src="figures/bridge.png"/&gt;
]

---

### Values: Hands-on Learning

We want workshop participants with varying levels of expertise to leave with a
concrete understanding of the current theory and practice of data integration.

.pull-three-quarters-left[
1. We will curate a collection of problems, datasets, and methods for data
integration which can serve as a resource for the wider community.

1. We will create opportunities for participants to present short research demos
and talks.
]
.pull-three-quarters-right[
&lt;img src="figures/workbook.png"/&gt;
]

---

### Schedule

1. The program will span roughly three weeks with an initial tutorial period.

1. Following the tutorial, research talks will be broken up with discussion
sessions, participant-led demos, and lightning talks.

.center[
&lt;img src="figures/timeline.png" width=750/&gt;
]

---

### Tutorials: Biological Data Management and Visualization

**Kris Sankaran** will provide an introduction to managing and visualizing
multi-omics data using tools from the Bioconductor ecosystem.

.center[
&lt;img src="figures/visualization-1.png" width=800/&gt;
]

---

### Tutorials: Introduction to Regression Trees

**Wei-Yin Loh** will offer a course on tree-based methods and their application
to mixed data types that arise in biomedical settings.

.center[
&lt;img src="figures/guide_1.png" width=680/&gt;&lt;br/&gt;
&lt;span style="font-size: 18px;"&gt;
An example regression tree, from the tutorial [8].
&lt;span&gt;
]

---
### Tutorials: Introduction to Regression Trees

**Wei-Yin Loh** will offer a course on tree-based methods and their application
to mixed data types that arise in biomedical settings.

.center[
&lt;img src="figures/genie3_overview.png" width=720/&gt;

&lt;span style="font-size: 18px;"&gt;
Trees are often used for integration [9].
&lt;span&gt;
]

---

### Tutorials: Generative Models for Multi-omics

**Kris Sankaran** will review concepts from probabilistic modeling, evaluation,
and interpretation that are the basis for many data integration methods.

.center[
&lt;img src="figures/blocking_omics.gif" width=800/&gt;
]

---

### Tutorials: Case Studies from CPTAC

**Xiaoyu Song** will lead a tutorial describing how problems have been
formulated
and solved in her expeience with the Clinical Proteomic Tumor Analysis
Consortium (CPTAC).

.center[
&lt;img src="figures/CPTAC_overview.jpg" width=700/&gt;
]

---

### Tutorials: Case Studies from CPTAC

**Xiaoyu Song** will lead a tutorial describing how problems have been
formulated
and solved in her expeience with the Clinical Proteomic Tumor Analysis
Consortium (CPTAC).

.center[
&lt;img src="figures/CPTAC_multiomics.png" width=800/&gt;
]

---

### Talks Criteria

We will invite speakers based on complementary criteria,

.pull-three-quarters-left[
1. **Compelling Applications**: Work that shows what is possible by carefully analyzing modern data.

1. **Statistical Foundations**: Research that draws deeply from the statistical canon and highlights opportunities for the field.

1. **Creative Perspectives**: Studies that have discovered surprising connections across fields.
]

.pull-three-quarters-right[
  &lt;img src="figures/presentation.png"/&gt;
]

---

### Interaction

Beyond the featured talks, we will create opportunities for both formal and
informal interaction among all participants.

.pull-three-quarters-left[
1. Facilitated discussion
1. Lightning research talks
1. Data analysis demos 
]
.pull-three-quarters-right[
  &lt;img src="figures/speech-bubbles.png"/&gt;
]

We will curate resources from these activities on a website dedicated to the workshop.

---

### Why IMS-NUS?

1. **Interdisciplinary emphasis**: The workshop relies on an environment that
allows researchers from multiple backgrounds to feel welcome.

1. **Statistics reputation**: Unlike related workshops on integration, we
emphasize the role of statistics, and the IMS - NUS connection will help.

1. **Location**: The strength of both the statistics and bioinformatics
communities in Singapore will ensure that the workshop balances theoretical and applied perspectives.

.center[
&lt;img src="figures/nus_ims_full.png" width=550/&gt;
]

---

### Recruitment

The organizing team will recruit students, scientists, and researchers locally
and abroad who have the potential to benefit from one another's expertise.

1. University Campuses: National University of Singapore, Nanyang Technological University
1. Research Institutes: Genome Institute of Singapore, The Bioinformatics Institute
1. Medical Schools/Hospitals: Duke - NUS Medical School, NUS Medicine, LKC Medicine

---

### Conclusion

We believe that this topic has the potential both to demonstrate the value of
statistics to the wider scientific community and to bring exciting, new problems
into the conversation in statistics.

Thank you for your attention and we are happy to take any questions.

---

### References

[1] B. Efron. "Size, power and false discovery rates". En. In: _Ann. Stat._ 35.4 (Aug. 2007), pp. 1351-1377.

[2] S. Gao et al. "Sparse GCA and Thresholded Gradient Descent". In: _Journal of Machine Learning Research_ 24.135 (2023), pp. 1-61. URL:
[http://jmlr.org/papers/v24/21-0745.html](http://jmlr.org/papers/v24/21-0745.html).

[3] J. P. Gygi et al. "Integrated longitudinal multiomics study identifies immune programs associated with acute COVID-19 severity and mortality". In: _Journal of Clinical
Investigation_ 134.9 (May. 2024). ISSN: 1558-8238. DOI: [10.1172/jci176640](https://doi.org/10.1172%2Fjci176640). URL:
[http://dx.doi.org/10.1172/JCI176640](http://dx.doi.org/10.1172/JCI176640).

---

[1] V. A. Huynh-Thu et al. "Inferring regulatory networks from expression data using tree-based methods". En. In: _PLoS One_ 5.9 (Sep. 2010), p. e12776.

[2] W. Loh. _Classification and Regression Trees by Example_. &lt;https://pages.stat.wisc.edu/~loh/ims21.pdf&gt;. Tutorial at the 2021 Cuasal Inference with Big Data Workshop hosted by IMS
- NUS. [Accessed 28-06-2024]. 2021.

[3] G. J. Martínez-Colón et al. "SARS-CoV-2 infection drives an inflammatory response in human adipose tissue through infection of adipocytes and macrophages". En. In: _Sci. Transl.
Med._ 14.674 (Dec. 2022), p. eabm9151.

[4] M. Peng et al. "Integration and transfer learning of single-cell transcriptomes via cFIT". En. In: _Proc. Natl. Acad. Sci. U. S. A._ 118.10 (Mar. 2021).

[5] Wikimedia. _File:Lung cancer HAVCR2.jpg - Wikimedia Commons - commons.wikimedia.org_. &lt;https://commons.wikimedia.org/wiki/File:Lung_cancer_HAVCR2.jpg&gt;. [Accessed 28-06-2024].

---

[1] N. R. Zhang et al. "A modified Bayes information criterion with applications to the analysis of comparative genomic hybridization data". En. In: _Biometrics_ 63.1 (Mar. 2007), pp.
22-32.

---

### Figure Attribution

* presentation button by Sara from &lt;a href="https://thenounproject.com/browse/icons/term/presentation-button/" target="_blank" title="presentation button Icons"&gt;Noun Project&lt;/a&gt; (CC BY 3.0)
* discussion by Cuan Studio from &lt;a href="https://thenounproject.com/browse/icons/term/discussion/" target="_blank" title="discussion Icons"&gt;Noun Project&lt;/a&gt; (CC BY 3.0)
* workbook by john sapuomah from &lt;a href="https://thenounproject.com/browse/icons/term/workbook/" target="_blank" title="workbook Icons"&gt;Noun Project&lt;/a&gt; (CC BY 3.0)
* bridge by Edy Subiyanto from &lt;a href="https://thenounproject.com/browse/icons/term/bridge/" target="_blank" title="bridge Icons"&gt;Noun Project&lt;/a&gt; (CC BY 3.0)
* Computer by annisa luthfiasari from &lt;a href="https://thenounproject.com/browse/icons/term/computer/" target="_blank" title="Computer Icons"&gt;Noun Project&lt;/a&gt; (CC BY 3.0)
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
