---
title: "Visualization and Multitable Methods for the Microbiome"
author: "Kris Sankaran"
date: "`r Sys.Date()`"
output:
  ioslides_presentation:
    fig_caption: true
    slide_level: 2
bibliography: ["bibliography.bib"]
link-citations: yes
csl: modern-language-association-8th-edition.csl
---

<style type="text/css">
code.r{
  font-size: 8px;

}

slides > slide {
  height: 900px;
  width: 1200px
}

iframe{
  height: 400px;
  width: 700px;
}
</style>

```{r knitr, echo = FALSE}
library("knitr")
opts_chunk$set(cache = TRUE, fig.align = "left", out.width = "650px", echo = FALSE)
```

## Outline ##

### `treelapse` ###
- Context: `treelapse` (timelapse for trees) is a new data visualization package
  motivated by the study of microbial dynamics
- Case Studies: How does `treelapse` facilitate microbiome research?
    + Differential Abundance: Revisiting the pregnancy data
    + Differential Dynamics: New views of perturbation effects

### Multitable Methods ###
- Context: How can we understand the relationship between different aspects of a complex
  system, by jointly analyzing multiple sources of data?
- Case Studies:
    + Canonical Correspondence Analysis: Sharing a template pipeline
    + Sparse CCA: Another multitable view of the perturbation

# `treelapse`: Principles and Implementation

## Focus plus Context ##

* It is often possible to study the same data at multiple scales
	- Full network vs. neighborhood of a node
	- Full time series vs. short time window
* These scales can be juxtaposed by "focusing" on a few elements of
  interest while remaining "context"
* Interactivity can allow rapid transitions between alternative scales

```{r focuscontextexample, out.width = "450px", fig.align = "center", fig.cap = "An example of the focus plus context principle, from [@rao1994table]."}
include_graphics("figure/focus_context.png")
```

## Linking ##

* Alternative representations of the same samples can be placed
  side-by-side to suggest covariation
* Can be useful for studying more than two dimensions at a time
* Conditional probability and database query interpretations
	- What are values other variables, conditional on constraints for
      some of them

```{r linkingexample, out.width = "430px", fig.align = "center", fig.cap = "An example of linked brushing, from [@voigt2002extended]."}
include_graphics("figure/Ggobi04brushlink.png")
```

## Implementation ##

* `D3` (Data Driven Documents) is a javascript library designed for creating
  interactive data visualizations
    - Very flexible, but requires substantial time investment
* `R` is a statistical computing language, used to prepare, visualize, model
  data [@ihaka1996r]
    - More limited visualization options, but easy to use
* `htmlwidgets` builds a bridge between these tools, allowing users to create
  `D3` visualizations using `R` code [@vaidyanathan2014htmlwidgets]

```{r htmlwidgets, out.width = "400px", fig.align = "center"}
include_graphics("figure/htmlwidgets.png")
```

## Visualization Proposals ##

* Differential Abundance
	- DOI Tree: Apply the focus plus context principle to allow the user to
    navigate microbial abundances across different parts of the taxonomic tree
	- DOI Sankey: Modify DOI Trees to display abundances across several groups
* Differential Dynamics
	- Timebox Trees: Apply the linking principle to highlight subsets of the
    taxonomic tree associated with user-selected patterns of microbial dynamics
	- Treeboxes: The converse of Timebox Trees, for selections along the tree.
* [Introductory vignette](http://statweb.stanford.edu/~kriss1/treelapse_intro.html)

# `treelapse`: Case Studies #

## Differential Abundance and Preterm Births ##

* [@digiulio2015temporal] tracked the abundance of microbes in the
during vaginal microbiome during pregnancy
* Ideally, it would be possible to develop clear microbial signatures
associated with preterm births.
* Identified 5 main Community State Types (CSTs)
* [http://statweb.stanford.edu/~kriss1/pregnancy.html](http://statweb.stanford.edu/~kriss1/pregnancy.html)

## Comparing CSTs ##

```{r pretermcsts, out.width = "800px", fig.cap = "The increased diversity among samples in CST 4 is represented by the relatively larger contribution of red edges to branches outside of the Firmicutes."}
include_graphics("figure/preterm_csts.png")
```

## Comparing CSTs ##

```{r pretermcstslacto, fig.cap = "Zooming into the Lactobacillaceae family, we notice that the difference between the remaining four CSTs is related to which types of Lactobacillus are most prominent."}
include_graphics("figure/preterm_csts_lacto.png")
```

## Differential Dynamics across Perturbations ##

* It would be interesting to know how abundance trajectories respond to "flash
  flood" perturbations
* We embedded our visualizations in a shiny app [@chang2015shiny], to allow us
  to switch between individuals easily
* [https://krisrs1128.shinyapps.io/treelapse_cc/](https://krisrs1128.shinyapps.io/treelapse_cc/)

## Differential Dynamics between Bacteroidetes and Firmicutes ##

* Zooming into a one-month window around the perturbation, for Subject B
  exhibits a noticeable drop in abundance of Firmicutes
* On the other hand, the Bacteroidetes appear relatively resilient
* All series for phyla are highlighted, using the tree to create the selection,
  though Bacteroidetes and Firmicutes are far more abundant than the rest

```{r bacteroidetes-resilience, out.width = "550px", fig.cap = "Abundances across phyla, in a two month window around the perturbation."}
include_graphics("figure/bacteroidetes_firmicutes.png")
```

# Multitable Analysis: Methods and Applications ##

## Context ## 

* Standard analysis of the microbiome focuses on one type of measurement at a
  time (16S, metabolomic, metagenomic), but many interesting scientific
  questions ask how systems interact
* Multitable methods, originally developed in statistics, chemometrics, and
  ecology, can be adapted to address these questions
* We have established best practices in situations frequently encountered in the
  microbiome
  - Exploratory views for differential abundance
  - Model interpretation after perturbation classification

## Multitable Exploration: Canonical Correspondence Analysis ##
* CCpnA is an ordination method that incorporates supplementary data [@ter1986canonical]
    - Coordinates are based on both microbial counts and sample characteristics
    - This is an alternative to applying PCA to microbial counts alone
* Template code for differential abundance analysis is published in
  [@callahan2016bioconductor]
* While a useful tool, CCpnA leaves certain microbiome applications out of reach
    - Can only handle two tables at a time
    - Can suffer in higher-dimensional settings

```{r correspondence, out.width = "725px", fig.cap = "An application of CCpnA to a differential abundance study, from [@callahan2016bioconductor].", fig.align = "center"}
include_graphics("figure/ccpna.gif")
```

## Multitable Model Interpretation ##

* How can we ordinate multiple sources of data, in a way that accentuates
  perturbation-specific variation?
* Idea: Select features across different sources using sparse classification
  methods, and perform multitable ordination on the results
  - By classifying samples as perturbed or not, only those features related to
    the perturbation will be selected
  - This is a meta-algorithm, into which many alternative selection and
    ordination methods can be input
* Multitable exploratory techniques can be used to better interpret the results
  from a formal model

## Case Study: Perturbation Data ##

* Microbes, genes, and metabolites are jointly used to classify samples within a
  single subject as perturbed, using sparse Linear Discriminant Analysis
  [@clemmensen2011sparse]
* Then the results are ordinated with Multiple Factor
  Analysis (MFA) [@le2008factominer]
    - The left panel provides sample scores and suggests that the selected
      features are strongly associated with perturbation state
    - The right panel summarizes relationships between selected features across the
      three data sources

```{r, out.width = "375px", fig.align = "center", fig.show = "hold", fig.cap = "The multitable model interpretation approach to perturbation analysis, for AAI."}
include_graphics(c("figure/mfa_scores.png", "figure/mfa_factors.png"))
```

## Case Study: Perturbation Data ##

* Alternatively, we can select features from a single table, and then apply
  sparse dimensionality reduction to the remaining tables
* Here, a subset of microbes is selected using sLDA, then sparse Canonical
  Correlation Analysis (sCCA) is applied to identify associated metagenomic
  functions
    - The left panel provides sample scores, relating samples based on similarity in microbial and metagenomic signatures
    - The right two panels describe the microbes and genes used in constructing these associations

```{r, out.width = "320px", fig.align = "center", fig.show = "hold", fig.cap = "sCCA can be used to perform unsupervised selection after an initial modeling step."}
include_graphics(c("figure/cca_xy_scores-1.png", "figure/cca-plots-1.png", "figure/cca-go-terms-1.png"))
```

## Conclusions ##

* Analysis of microbial perturbations has motivated our development and
  adaptation of visualization and multitable methods
* With `treelapse` we have created a new, openly available visualization package
* Through several case studies, we have begun to establish best practices in
  multitable data analysis of microbial perturbation
    - Blending modeling and exploratory techniques
    - Leveraging sparsity in both supervised and unsupervised analysis

## References ##
