---
title: "Meditation and the Microbiome"
author: "Xinran Miao, Hanying Jiang, Kris Sankaran"
date: "April 7, 2022"
output:
  xaringan::moon_reader:
    css: ["default", "css/xaringan-themer.css"]
    lib_dir: libs
    self_contained: false
    nature:
      highlightStyle: github
      highlightLines: true
      ratio: "16:9"
    seal: false
---

class: top, my_title

$\def\Dir{\text{Dir}}$
$\def\Gsn{\mathcal{N}}$
$\def\Mult{\text{Mult}}$
$\def\*#1{\mathbf{#1}}$
$\def\m#1{\boldsymbol{#1}}$
$\def\Unif{\text{Unif}}$
$\def\win{\tilde{w}_{\text{in}}}$
$\def\reals{\mathbb{R}}$
$\newcommand{\wout}{\tilde w_{\text{out}}}$
$\def\naturals{\mathbb{N}} % Natural numbers$
$\def\Ber{\text{Bernoulli}}$

```{r, echo = FALSE, message = FALSE, warning = FALSE}
library(knitr)
library(tidyverse)
library(RefManageR)
opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = T, dpi = 200, fig.align = "center", fig.width = 6, fig.height = 3)
min_theme <- theme_minimal() + 
  theme(
    panel.grid.minor = element_blank(),
    panel.background = element_rect(fill = "#f7f7f7"),
    panel.border = element_rect(fill = NA, color = "#0c0c0c", size = 0.6),
    axis.text = element_text(size = 14),
    strip.text = element_text(size = 16),
    axis.title = element_text(size = 16),
    legend.position = "bottom"
  )
theme_set(min_theme)
BibOptions(
  check.entries = FALSE, 
  bib.style = "numeric", 
  cite.style = "numeric",
  style = "markdown",
  hyperlink = FALSE, 
  dashed = FALSE
)
bib <- ReadBib("20220408.bib")
```


## Meditation and the Microbiome

.pull-left[
Xinran Miao, Hanying Jiang, and Kris Sankaran <br>
UW Madison
]

.pull-right[
Holmes Lab <br>
April 7, 2022
]

---

### Setting

.pull-left[
* There was a time when many academics did not consider meditation a “serious” topic for research
* UW-Madison was one of the few institutions willing to support this work
* Richie Davidson’s lab here has helped change the discourse on this topic
]

.pull-right[
```{r, echo = FALSE, out.width = 400}
include_graphics("figure/fmri.png")
```
]

---

### Meditation meets Genomics

Given all the changes that meditation induces on physiological processes, it is natural to ask whether there is a relationship with the microbiome.

This is not so far fetched,
* There are documented relationships with anxiety and depression `r Citep(bib, c("winter2018gut", "peirce2019role", "dash2015gut"))`
* The mechanisms may be related to the Gut-Brain Axis `r Citep(bib, c("foster2013gut", "carabotti2015gut", "mayer2015gut"))`

---

### Study Population and Design

```{r, echo = FALSE}
sample_data <- read_csv("../../analysis/data/derived_data/metadata/cleaned_data.csv") %>%
  mutate(subject = as.factor(subject))
```


The Center for Healthy Minds (CHM) recruited police officers into the study,
since they have high stress jobs. The treatment group received meditation
training just before the second timepoint.

```{r, echo = FALSE, fig.width = 8, fig.height = 6, out.width = 520}
sample_data %>%
  mutate(intervention = case_when(
      group == "WLC" ~ "Control",
      group == "MBRT" & timepoint == 1 ~ "Pre-Intervention",
      group == "MBRT" & timepoint == 2 ~ "Intervention",
      group == "MBRT" & timepoint == 3 ~ "Post-Intervention",
    ),
    intervention = factor(intervention, levels = c("Control", "Pre-Intervention", "Intervention", "Post-Intervention"))
  ) %>%
  ggplot() +
  geom_point(aes(technical.replicate, subject, col = intervention), size = 2) +
  facet_grid(group ~ timepoint, scales = "free") +
  scale_color_manual(values = c("#D99152", "#F2C572", "#423F8C", "#BF4B75")) +
  theme(
    legend.text = element_text(size = 14),
    axis.text.y = element_text(size = 9), 
    legend.title = element_blank()
  )
```

---

### Available Data

* 16S sequencing, up to 4 technical replicates for each sample
* Static demographic + time-varying diet and physical health data
* Time-varying mental health survey responses
* Consistency of meditation practice following treatment

---

### Preliminary Views

The 16S data are very high-quality.

```{r}
include_graphics("figure/dada_preprocess.png")
```

---

### Preliminary Views

Sequencing depth is not consistent across the three timepoints.
  - Working with absolute abundances leads to spurious treatment effect
  - We transform to relative abundances in Model 1

```{r}
metadata <- read.csv("../data/derived_data/metadata/cleaned_data.csv") %>%
  mutate(replicate = technical.replicate)
totals <- read.csv("../data/derived_data/taxa_level/joined_data_Genus.csv") %>%
  dplyr::filter(!is.na(group)) %>%
  dplyr::mutate(subject = as.numeric(subject))  %>%
  dplyr::select(filename, group, timepoint, Blautia:Anaerotruncus) %>%
  pivot_longer(Blautia:Anaerotruncus, names_to = "taxa") %>%
  group_by(filename, group, timepoint) %>%
  summarise(total = sum(value)) %>%
  dplyr::mutate(intervention = case_when(
      group == "WLC" ~ "Control",
      group == "MBRT" & timepoint == 1 ~ "Pre-Intervention",
      group == "MBRT" & timepoint == 2 ~ "Intervention",
      group == "MBRT" & timepoint == 3 ~ "Post-Intervention",
    ),
    intervention = factor(intervention, levels = c("Control", "Pre-Intervention", "Intervention", "Post-Intervention"))
  )
```


```{r, fig.width = 8, fig.height = 4, out.width = 750}
ggplot(totals) +
  geom_histogram(
    aes(total, fill = as.factor(intervention)),
    position = "identity", alpha = 0.8
    ) +
  scale_fill_manual(values = c("#D99152", "#F2C572", "#423F8C", "#BF4B75")) +
  scale_y_continuous(expand = c(0, 0, 0.1, 0)) +
  facet_grid(timepoint ~ group) +
  theme(
   legend.text = element_text(size = 14),
   axis.text.y = element_text(size = 10), 
   legend.title = element_blank()
  ) +
  labs(x = "Sequencing Depth")
```

---

### Approach

* For each taxonomic level,
	- Model 1: Mixed-effect model on transformed responses
	- Model 2: Multinomial regression model on raw counts
* Compare model similarities / differences, evaluate goodness-of-fit, and refine
accordingly

---

# Model 1: Linear Mixed-Effects Model

---

### Main Idea

* Treatment: We view the short-term effect of meditation as a fixed treatment effect
* Variation: We add subject-specific random effects
* Response: Log transformed relative abundances
  - Transformation will not be needed in model 2
* We fit a separate model for each taxa at genus level,
  - No pooling: We ignore possible correlations across taxa.



---
### Notation

* $i, j, k$ index subjects, replicates, and timepoints, respectively
* $y_{ijk}$: log-transformed relative abundance
* $x_{ijk}$: scaled and centered anxiety scores
* Short-term meditation-effect indicators,

\begin{align*}
    t_{ijk} = \begin{cases}
    1, & \mathrm{if\ }k=2\ \mathrm{and\ sample\ }i\ \mathrm{in \ the\ treatment\ group}\\
    0, & \mathrm{otherwise}
    \end{cases}
\end{align*}

---

### Model (Math)

For each species, we fit,

\begin{align}
y_{ijk} =  \beta_0 + \beta_1 x_{ijk} + \beta_2 t_{ijk} + e_i +\varepsilon_{ijk}
\end{align}

with different subject-to-subject effects,

\begin{align}
\varepsilon_{ijk} \sim \Gsn(0, \sigma^2) \\
e_{i} \sim \Gsn\left(0, \sigma_i^2\right)
\end{align}

In code, this is the model,

```{r, eval = FALSE, echo = TRUE}
stan_glmer( y ~ (1 | subject) + trt + promis_anx, data = dat)
```

---

### Model (Interpretation)

| Notation | Interpretation |
|---|------|
| $\beta_0$ | The overall fixed intercept |
| $\beta_1$ |  If we increase anxiety score by one unit, how will the log-transformed relative abundance change, all else held equal?|
| $\beta_2$ |  Treatment effect: If we apply the intervention, i.e. $t_{ijk}$ changed from $0$ to $1$, how will the log-transformed relative abundance change, all else held equal? |
| $\sigma^2$ |  The variation of relative abundance in all samples. Represents noise resulting from unmeasured factors across all subjects. |
| $\sigma_i^2$ |  A subject $i$ specific variance parameter, used to account for differential noise levels across subjects.

---

### Filtering

* We only fit to taxa that appeared in at least 20% of the samples
  - $134 \to 55$ genera total
* Examining residuals, it seems necessary to screen for outliers
  - We remove samples whose residuals lie outside the 1.5 IQR for each taxa
  - In each model, numer of outliers is usually fewer than 3.

```{r, out.width = 680}
include_graphics("figure/diag_original.png")
```
---

### Evaluation

* Removing outliers helps, but some residuals are still skewed.

```{r, out.width = 800}
include_graphics("figure/remove_outlier.png")
```
---
### Evaluation

* Subject-to-subject effects can have very different distributions across taxa
```{r, out.width = 800}
include_graphics("figure/random_effects.png")
```
---
### Evaluation

By fitting models taxon-by-taxon, we are able to preserve relative abundance profiles.

```{r, out.width = 600}
include_graphics("figure/sig_prob_mixed.png")
```

---

### Evaluation
We cannot capture the dispersion well, expecially when data are highly skewed by zero / non-zeros.
```{r, out.width = 680}
include_graphics("https://raw.githubusercontent.com/krisrs1128/CHMMicrobiome/main/analysis/taxa_level_analysis/mixed_effects/figures/Genus/Genus_violin_sig_taxa.png?token=GHSAT0AAAAAABRPZ7UIOBOPK2GHUCIRZDXGYSXSAMA")
```





---

### Interpretation


CI’s for $\beta_{2}$ can guide our search for taxa that are impacted by meditation.

```{r, out.width = 600}
include_graphics("https://raw.githubusercontent.com/krisrs1128/CHMMicrobiome/main/analysis/taxa_level_analysis/mixed_effects/figures/Genus/Genus_coef.png?token=GHSAT0AAAAAABRPZ7UIHZZWEABVL7JVZKB4YSXROUQ")
```

---
### Possible next step
* The above model cannot address excessive zeros.
* We can introduce a parameter $p$ as the probability of having zero count, and fit a hurdle model:

\begin{align*}
 &z_{ijk}\sim\Ber{(p)}\\
    &y_{ijk} = \begin{cases}
    \beta_0 + \beta_1 x_{ijk} + \beta_2 t_{ijk} + e_i +\varepsilon_{ijk}, & z_{ijk}=0\\
    0, & z_{ijk}=1
    \end{cases}
\end{align*}

---

# Model 2: Multinomial Model

---

### Motivation

We can potentially improve on mixed-effects by directly modeling count profiles
- Avoids relative abundance transformation
- Partial pooling across related taxa

---

### Notation

* $N_{i}$: Total number of reads in sample $i$
* $y_{i} \in \naturals^{D}$: Taxa count profile for sample $i$
* $x_{i} \in \naturals^{K}$: Covariates for sample $i$, including
    + the anxiety scores
    + a short-term meditation-effect indicator
    + the one-hot encoding of subject
---

### Model (Math)

\begin{align*}
y_{i} &\sim \Mult\left(N_{i}, p_{i}\right) \\
\theta_{i} &=  \alpha + x_{i}^T \beta\\
p_{i} &= \text{Softmax}\left(\theta_{i}\right) \\
\beta_{i,j} &\overset{\mathrm{iid}}{\sim} \Gsn\left(0, \sigma^2 \right) \\
\alpha_{j} &\overset{\mathrm{iid}}{\sim} \Gsn\left(0, \sigma^2 \right)
\end{align*}

Notice that effects $\beta$ are $K\times D$ matrix, with one column per taxa.

---

### Model (Code)

```{r, echo = TRUE, eval = FALSE}
model {
  to_vector(alpha) ~ normal(0, sigma_beta);
  to_vector(beta) ~ normal(0, sigma_beta);
  for (i in 1:n){
      vector[K] theta = alpha + (X[i] * beta)';
      vector[K] P = softmax(theta);
      Y[i] ~ multinomial(P);
  }
}
```
---
### Informative vs. Uninformative Prior
We consider two types of prior.
.pull-left[
Informative prior:
```{r}
include_graphics("figure/simulated_distribution_smallprior.png")
```
]
.pull-right[
Uninformative prior:
```{r}
include_graphics("figure/simulated_distribution_largeprior.png")
```
]
---
### Informative vs. Uninformative Prior
With the informative prior, the sampling is more robust and converges quickly.
.pull-left[
Informative prior:
```{r}
include_graphics("https://raw.githubusercontent.com/krisrs1128/CHMMicrobiome/main/analysis/notes/figure/sampling_trace_informative.png?token=GHSAT0AAAAAABQKHWMBVG3B56433MIUNEAKYSXQQ5A")
```
]
.pull-right[
Uninformative prior:
```{r}
include_graphics("https://raw.githubusercontent.com/krisrs1128/CHMMicrobiome/main/analysis/notes/figure/sampling_trace_uninformative.png?token=GHSAT0AAAAAABQKHWMBSQ56OIAB7ORRKSO4YSXQRNA")
```
]
---
### Evaluation
Relative abundances are captured well.
```{r, out.width = 700}
include_graphics("https://raw.githubusercontent.com/krisrs1128/CHMMicrobiome/main/analysis/taxa_level_analysis/softmax_regression/analysis/figures/Genus/informative/P_vs_true_abundance_aggregated/figure1.png?token=GHSAT0AAAAAABQKHWMB6XGTASTJJU7SDL5CYSXSOCQ")
```
---
### Evaluation
* While the model captures relative abundance levels well, it struggles with skewness / overdispersion.
* This results in simulated treatment effects that look artificially strong.
```{r, out.width = 1000}
include_graphics("figure/fitted_Phylum.png")
```
---
### Evaluation
Correlation between taxa is captured surprisingly well across taxa
- Likely due to variation in $p_i$ across samples
.pull-left[
Real:
```{r}
include_graphics("https://raw.githubusercontent.com/krisrs1128/CHMMicrobiome/main/analysis/taxa_level_analysis/softmax_regression/analysis/figures/Genus/informative/correlation/true_correlation.png?token=GHSAT0AAAAAABQKHWMAUFBRF2NZBFMAXKR2YSXSPUA")
```
]
.pull-right[
Simulated:
```{r}
include_graphics("https://raw.githubusercontent.com/krisrs1128/CHMMicrobiome/main/analysis/taxa_level_analysis/softmax_regression/analysis/figures/Genus/informative/correlation/simulated_correlation.png?token=GHSAT0AAAAAABQKHWMASA4HTD7FK5JXAFTEYSXSQAA")
```
]
---
# Interpretation
We can also define p-value here.
\begin{align*}
z &= \frac{\hat{\beta}}{\hat{\sigma}_{\beta}}\\
p &= 2 \times\Phi(-\left\lvert z \right\rvert)
\end{align*}
---
# Interpretation
Use the defined p-value to find the taxa significantly influenced by treatment: 
.pull-left[
```{r}
include_graphics("figure/example_of_sig.png")
```
]
.pull-right[
```{r}
include_graphics("figure/example2_of_sig.png")
```
]
---
# Discussion
* This model can capture both structural and sampling zeros.
* Even though convergence speed differs, the fitted model is robust to the choice prior.
* Effectively models relative abundance of taxa.
* Better testing is of further interest.
---
### Comparing Models
* We can study which effects are shared / specific for the two models
* Discrepancies could be explained by...
  - Outlier preprocessing was different across the models
  - Log transformation could place greater emphasis on samples with lower
  total read counts
* Limited interpretation of multinomial model
  
---
### Model-Specific, Significant Positive Effects
```{r, out.width=810}
include_graphics("https://raw.githubusercontent.com/krisrs1128/CHMMicrobiome/main/analysis/taxa_level_analysis/softmax_regression/diagnostic/comparison/Genus/violin_pos_all.png?token=GHSAT0AAAAAABMZE4BKXNCFIDRO52INAQVOYSU6WBQ")
```
---
### Model-Specific, Significant Negative Effects
```{r, out.width=810}
include_graphics("https://raw.githubusercontent.com/krisrs1128/CHMMicrobiome/main/analysis/taxa_level_analysis/softmax_regression/diagnostic/comparison/Genus/violin_neg_all.png?token=GHSAT0AAAAAABMZE4BK45OQSDVQ742XCUJ6YSU66UA")
```
---
### Shared, Significant Negative Effects
```{r, out.width = 800}
include_graphics("https://raw.githubusercontent.com/krisrs1128/CHMMicrobiome/main/analysis/taxa_level_analysis/softmax_regression/diagnostic/comparison/Genus/violin_neg_Both.png?token=GHSAT0AAAAAABMZE4BL7C3TJB5WSZ5YJ7Q6YSU63JA")
```
---
### Potential problem with the p-value in multinomial model
$0$ probably cannot be the criterion to determine whether effect exists/whether the effect is positive or negative.
```{r, out.height= 400, out.width=600}
include_graphics("figure/limitation_p.png")
```
---
### Tentative Conclusions
* Among the species we find significant, a few have been previously documented in the literature around the microbiome and mental health.
* It would be interesting if meditation is having the same effect on the microbiome as anti-depressants.
* But we haven’t shared this with our collaborators yet, so there might be other interpretations.
---
### Potential Next Steps
* Better capturing skewness / overdispersion
* Incorporating additional survey information
  - Can control for within-subject differences other than anxiety
* Estimating false discovery rates
  - We haven’t accounted for multiple testing
* Interactively comparing species estimates
---
### References
```{r, results = "asis"}
PrintBibliography(bib)
```
