---
title: "Visualization of Hierarchically Structured Data"
author: "Kris Sankaran"
date: "`r Sys.Date()`"
output:
  ioslides_presentation:
    css: talk.css
    self_contained: no
bibliography: ["bibliography.bib"]
csl: proceedings-of-the-royal-society-a.csl
---

```{r knitr, echo = FALSE}
library("knitr")
opts_chunk$set(cache = TRUE, fig.align = "center", echo = FALSE)
read_chunk("20170512_talk.R")
```

## Outline

### Context
* What is the microbiome?
* Review literature on high-dimensional data visualization
    + Focus + Context [@herman2000graph, @heer2004doitrees, @hochheiser2004dynamic]
    + Linking [@buja1996interactive]
* Description of our R package, treelapse (timelapse for trees)

### Case Studies
* Exploratory analysis: Antibiotics and the microbiome
* Modeling: CART and hierarchical clustering
* Inference: Known structure between hypotheses

# Microbiome Background

## Microbiome Analysis

* Scientific Motivation
    + Characterize variation in bacterial abundance profiles across experimental
      conditions [@human2012structure]
    + Ecology meets medicine: Behavior of bacterial communities can have
      medical implications

```{r, out.width = "436.67px", out.height = "275.275px", , fig.cap = "See [@diemart2015; @specter2012; @kolata2017]"}
include_graphics("figure/microbiome_montage.png")
```

## Data Sources

* Sequencing of the 16S marker gene gives counts of different sequence
  variants, across samples
* Typically 10 - 100 samples, 500 - 2000 bacteria
* Outside of bacterial counts, contextual data are available both about
  bacteria (taxonomic information) and samples (subject characteristics)

```{r, out.width = "723.8px", out.height = "154px"}
include_graphics("figure/microbiome_data.png")
```

## Motivating Statistical Problems

* (Tree-structured) Differential Abundance
	- Compare bacterial abundances across conditions, with emphasis on
    middling taxonomic orders
	- Identify the largest taxonomic subtree whose associated bacteria
    are differentially abundant

```{r, out.width = "243.33px", out.height = "260.8px"}
include_graphics("figure/differential_abundance.png")
```

## Motivating Statistical Problems

* (Tree-structured) Differential Dynamics
	- Describe changes in bacterial abundances, at the highest subtree
    where the pattern appears
	- Has an ecological flavor, with emphases on role of niches
    and environmental changes

```{r, out.width = "520px", out.height = "260.8px"}
include_graphics("figure/differential_dynamics.png")
```

# Data Visualization Background

## Visualization Principles: Focus and Context

* It is often possible to study the same data at multiple scales
	- Full network vs. neighborhood of a node
	- Full time series vs. short time window
* These scales can be juxtaposed by "focusing" on a few elements of
  interest while retaining "context"
* Interactivity can allow rapid transitions between alternative scales

## Degree-of-Interest (DOI) Trees ##

* An application of the "focus + context" idea to tree structured data
* Should be able to easily transition between visualization at different scales
  (e.g., Species vs. Phylum)

<iframe src="http://www.youtube.com/embed/RTQ0N4QY0yc?html5=1" style="height: 350px;"></iframe>

## DOI Calculation ##

* The main idea is to define a DOI distribution on nodes of
the tree [@heer2004doitrees]
* Then, the display is determined by recursively trimming subtrees with low
interest
* Each time a node is clicked, the DOI distribution is recalculated

```{r, doi_fig, out.width = "459.5px", out.height = "166.5px", echo = FALSE}
include_graphics("doi_distn.png")
```

## TimeBoxes ##

* You can "query" a collection of time series visually by sketching its shape
  [@hochheiser2004dynamic]

<iframe src="http://www.youtube.com/embed/VWx1TMcrb74?html5=1" style="height: 350px;"></iframe>

## Visualization Principles: Linking ##

* Alternative representations of the same samples can be placed side-by-side to
  suggest covariation [@buja1996interactive, @becker1987brushing]
* Can be useful for studying more than two dimensions at a time
* Conditional probability and database query interpretations - What are values
  other variables, conditional on constraints for some of them

<a href="https://bl.ocks.org/mbostock/4063663">
  <img src="d3_block.png" style="margin: auto; display: block; height: 250px;">
</a>

# Treelapse Package

## Visualization Proposals ##

* Differential Abundance
	- DOI Tree: Standard DOI implementation, trivially extended to show node
    abundances.
	- DOI Sankey: Modify DOI Trees to multiple groups

```{r, out.width = "588.67px", out.height = "204.47px"}
include_graphics("figure/annotated_preterm_csts.png")
```

##  Visualization Proposals

* Differential Dynamics
	- Timebox Trees: Link timeboxes and tree, with selections over time series
	- Treeboxes: Link timeboxes and tree, with selections over the tree

```{r out.width = "674px", out.height = "277.5px"}
include_graphics("figure/annotated_antibiotic_overview.png")
```

## Implementation Tricks

* htmlwidgets: To call javascript code from R, we use the htmlwidgets package
* Voronoi mouseover: Drawing mouseover labels at the node closest to the current
  mouse position facilitates UI
* Search and autocomplete: This simplifies the search for apriori interesting
  taxa

```{r out.width = "595.625px", out.height = "133.125px"}
include_graphics("figure/voronoi.png")
```

# Case Studies

## Bacterial Dynamics and Antibiotic Time Courses

* Revisiting the study [@dethlefsen2008pervasive]
* Study Goal: What is the response of the microbiome following antibiotics,
  in comparison to natural long-term variation? ["Flash Flood" analogy]
* Data: 3 subjects, across ~ 50 timepoints, with two antibiotic time courses
  introduced in between
* We study one subject at a time, and filter away less abundant microbes
* [http://statweb.stanford.edu/~kriss1/antibiotic.html](http://statweb.stanford.edu/~kriss1/antibiotic.html)


```{r, out.width = "674.5px", out.height = "253.5px"}
include_graphics("figure/antibiotic_overview.png")
```

## Bacteroidetes are unaffected

* We can sketch timeboxes to identify series with no change
* After some study, it becomes clear that one phylum does not seem to be
  strongly affected by the first antibiotic time course

```{r, out.width = "674.5px", out.height = "253.5px"}
include_graphics("figure/antibiotic_bacteroidetes.png")
```

## Differential Response among Firmicutes

* Drawing boxes over the tree and searching for different taxonomic orders
  suggests differential responses between the Lachnospiraceae and
  Ruminococcaceae
* Note that we are focusing on the time window surrounding the second antibiotic
  time course

```{r, out.width = "674.5px", out.height = "253.5px"}
include_graphics("figure/antibiotic_firmicutes.png")
```

## Averages vs. Sums on Tree

* We have effectively only been studying sums across large taxonomic groups
* To understand finer-grained taxa, we can either use the pan-zoom widget to
  emphasize lower scales, or consider averages up the tree instead of sums
* We use the second approach to inspect recovery among Ruminococcaceae

```{r, out.width = "674.5px", out.height = "253.5px"}
include_graphics("figure/ruminococcus.png")
```

## Increases in Verrumicrobiae

* The averaged view from before can also be used to identify series that
  increase during the antibiotic treatment
* The average abundance across the Verrumicrobiae seems to increase after the
  antibiotic treatment

```{r, out.width = "674.5px", out.height = "253.5px"}
include_graphics("figure/verrucomicrobiae.png")
```

## Variation in Bikesharing Demand

* How does demand for bikesharing services vary over time?
* We construct a collection of time series by splitting one long series into
  many daily ones
* Two-table structure can be used to construct a regression tree
  	- Two tables are bike demand and supplementary weather / date features
  	- Relate the two tables in an interpretable tree structure using a
      (slightly overgrown) regression tree that predicts 8AM demand
* [http://statweb.stanford.edu/~kriss1/bikesharing.html](http://statweb.stanford.edu/~kriss1/bikesharing.html)

## First split: Weekday vs. Weekend Effect

* The two peaks at rush hour distinguish weekday series from the rest
* Unlike weekday demand, weekend demand is unimodal
* The few weekday series with unimodal series are associated with holidays

```{r out.width = "450px", out.height = "170px", fig.align = "default", fig.show = "hold"}
include_graphics(c("figure/working.png", "figure/weekend.png"))
```

## Lower-Order Effects: 2012 vs. 2011

* We can interactively study deeper splits that might be harder to interpret
  otherwise (at least, if we care about hours other than 8AM)
* For example, the second level split indicates that weekday demand is larger in
  2012 than it was in 2011
```{r out.width = "675px", out.height = "255px", fig.show = "hold"}
include_graphics("figure/weekday_2011.png")
```

## Brushing Time Windows

* In contrast to these general questions about daily demand, we could ask
  granular questions about specific time windows
* E.g., what characterizes days on which there is larger than average demand
  after midnight?
* These series seems associated with the warm-weekend split, which seems
  reasonable in retrospect

```{r out.width = "449px", out.height = "169.7px", fig.show = "hold"}
include_graphics("figure/warm_weekend.png")
```

## DOI on the CART Tree

* Group samples according to their quintile of 8AM demand and then count the
  abundance of the groups flowing down different branches.
* If the model is a good fit, then the quintiles separate after only a few steps
  down the regression tree

```{r out.width = "607px", out.height = "284px", fig.show = "hold"}
include_graphics("figure/bike_sankey.png")
```

## Interaction and Hierarchical Clustering

* Instead of time series, we can consider collections of parallel coordinates [@inselberg1990parallel]
* The Global Patterns data
    - How do the microbiomes compare across ecological environments (freshwater, soil, human gut, ...)?
    - Data are 26 samples across 9 ecological environments
    - 19216 bacteria each, though we filter to 500 most variable
* We hierarchically cluster the samples using complete linkage and the UniFrac
  distance [@lozupone2005unifrac]
* [http://statweb.stanford.edu/~kriss1/global_patterns.html]

## Comparing Clusters

* We can compare two subclusters after zooming to a few of the bacteria
  that distinguish between them
* After revisiting the original data, becomes clear that samples on the left
  come from freshwater samples, while those on the right come from soil and skin
   - Also find that species with high abundances at left from Taxonomic
     order Oceanospirallales

```{r, out.width = "338px", out.height = "295px", fig.align = "default", fig.show = "hold"}
include_graphics(c("figure/gp_cluster1.png", "figure/gp_cluster2.png"))
```

## Inspecting Confirmatory Analysis

* Can we use interactive visualization to better understand results from formal
  statistical techniques?
    - Opposed to only ever exploring raw data
* Example: How are the microbiomes of old and young mice different
  [@callahan2016bioconductor]?
* We would like to identify associations at multiple taxonomic resolutions

## Ideal Visualization

* To natural approaches do not fit into the treelapse abstraction
    - We could display curves of $p$-values, as in manhattan plots
    - We could display the raw data, shaded in by significance

```{r out.width = "647.86px", out.height = "314.29px"}
include_graphics("figure/ideal_pval.png")
```

## Alternative: Comparing Methods

* The $y$-axis gives negative log $p$-values (both raw and adjusted) after
  testing all leaf and aggregated node-level data
* Two clusters of bacteria appear significantly associated, across all methods
* The Sidak one-step and structSSI procedures seem to have lower power, even in
  comparison with FWER-controlling methods [@dudoit2008multiple,
  @sankaran2014structssi]

```{r, out.width = "451px", out.height = "319px"}
include_graphics("figure/structssi.png")
```

## Conclusion ##

* We've implemented some old but good visualization ideas in a new R package
* We've described the practical usefulness of these principles in real-world
  data analysis
* Future work
    - Alternative structures: Similar ideas could be useful in the analysis
        of networks or spectra in the microbiome context
    - Visualizing uncertainty: More than simply delivering "insights", it's our role as
       statisticians to quantify degrees of uncertainty in them

## References {#references}
