<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Semisynthetic Simulation for Biological Data Analysis</title>
    <meta charset="utf-8" />
    <meta name="author" content="Kris Sankaran" />
    <script src="libs/header-attrs-2.27/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">


class: title



&lt;div id="title"&gt;
Semisynthetic Simulation for Biological Data Analysis
&lt;/div&gt;
&lt;div id="under_title"&gt;
Session 3: Integrative Analysis
&lt;/div&gt;

&lt;div id="subtitle"&gt;
Kris Sankaran &lt;br/&gt;
25 | June | 2024 &lt;br/&gt;
Lab: &lt;a href="https://go.wisc.edu/pgb8nl"&gt;go.wisc.edu/pgb8nl&lt;/a&gt; &lt;br/&gt;
&lt;/div&gt;

&lt;div id="subtitle_right"&gt;
Melbourne Integrative Genomics&lt;br/&gt;
Slides: &lt;a href="https://go.wisc.edu/54tmr9"&gt;go.wisc.edu/54tmr9&lt;/a&gt;&lt;br/&gt;
Code: &lt;a href="https://go.wisc.edu/yd3yz9"&gt;go.wisc.edu/yd3yz9&lt;/a&gt;
&lt;/div&gt;

---

### Learning Outcomes:

1. Provide examples of horizontal and vertical integration problems and identify
issues that can complicate their analysis.
1. Discuss how simulation can help in assessing the validity and power in
horizontal and vertical integration workflows.
1. Design simulators that respect batch or multi-omics structure to support
integration use cases.

---

class: middle

.center[
## Horizontal Integration
]

---

### Motivation

1. We can learn a lot by paying attention to as many relevant datasets as
possible. We shouldn't pretend that we haven't seen related data before!

1. Unfortunately, technical factors can cause systematic differences between
groups of samples. These are often called "batch effects."

.center[
&lt;img src="figures/horizontal_integration.png" width=600/&gt;
]

---

### Sources of Batch Variation

These kind of unwanted variation can result from:

* Variation in experimental or sequencing protocols.

* Differences in library size.

* Variation in dataset storage methods.

* Uninteresting, but real, biological variation: situation in cell cycle,
differing proportions of dead cells.

.center[
]

---

### General Approaches

Many methods have been proposed to do this. They mostly fall into one of two
categories,

1. Correct: Normalize the data before any downstream analysis.
1. Account: Directly model batch-to-batch variation in the downstream analysis.


|  | Pros | Cons |
|--|---|---|
| Correct | Generality. Allows for any downstream analysis. | Can introduce its own artifacts. |
| Account | Cohesion. Only one method needed for everything.  | Time consuming. Methods become "to big to fail." |

---

### Controls

A powerful idea in correction methods is to make proper use of controls:

* Replication: At specific stages of analysis, you can create replicates to clarify the nature of subsequent variation.
* Negative Controls: The experiment can be designed so that some features will not vary with interesting biological variation.

---

### Controls

Effective replication and controls can help distinguish between upstream
biological variation (from the original sample) and technical variation
introduced later (during sample preparation and sequencing).

.center[
&lt;img src="figures/variation_stages.png" width=900/&gt;
]

---

### Challenges

Correction needs to be applied with care. It's easy for problems to arise.

.pull-left[
Overcorrection: Real biological variation can be masked by overly aggressive
batch effect correction methods.
]

.pull-right[
&lt;img src="figures/cell-anova.png" width=500/&gt;
Figure from (Zhang et al., 2023).
]

---

### Challenges

Correction needs to be applied with care. It's easy for problems to arise.

.pull-left[
Miscalibration: Plugging in batch-effect corrected outputs into differential
analysis workflows can lead to miscalibrated `\(p\)`-values. 
]

.pull-right[
&lt;img src="figures/correction-1.png"/&gt;
Figure from (Tonkin-Hill, 2023).  See also (Nygaard et al., 2016).
]

---

### Challenges

Correction needs to be applied with care. It's easy for problems to arise.

.pull-left[
Miscalibration: Plugging in batch-effect corrected outputs into differential
analysis workflows can lead to miscalibrated `\(p\)`-values. 
]

.pull-right[
&lt;img src="figures/correction-2.png"/&gt;
Figure from (Tonkin-Hill, 2023).  See also (Nygaard et al., 2016).
]

---

### Challenges

Correction needs to be applied with care. It's easy for problems to arise.

.pull-left[
Miscalibration: Plugging in batch-effect corrected outputs into differential
analysis workflows can lead to miscalibrated `\(p\)`-values. 
]

.pull-right[
&lt;img src="figures/correction-3.png"/&gt;
Figure from (Tonkin-Hill, 2023).  See also (Nygaard et al., 2016).
]

---

### Challenges

Correction needs to be applied with care. It's easy for problems to arise.

.pull-left[
Miscalibration: Plugging in batch-effect corrected outputs into differential
analysis workflows can lead to miscalibrated `\(p\)`-values. 
]

.pull-right[
&lt;img src="figures/correction-4.png"/&gt;
Figure from (Tonkin-Hill, 2023).  See also (Nygaard et al., 2016).
]

---

### Simulation

Simulation can give insight into the behavior of correction methods and their
influence on downstream analysis. 

1. Overcorrection: We can simulate possible forms of biological and batch
variation and benchmark using this as ground truth.

1. Miscalibration: We can create data with known null and nonnull genes and
calculate power and false discovery rates.

---

### Simulator Definition

We will assume that the batches are known and stored in the experiment's
`colData`. These will be allowed to influence estimation of GAMLSS parameters
like any other covariate.


``` r
simulator &lt;- setup_simulator(
  exper,
  ~ batch + ...,
  ...
) |&gt;
  estimate()
```

---

.center[
## Demo + Exercises
]

* Code repository: [go.wisc.edu/o5sn6w](https://go.wisc.edu/o5sn6w)
* Exercise solutions: [go.wisc.edu/yd3yz9](https://go.wisc.edu/yd3yz9)
* Live Link: [go.wisc.edu/wj5665](https://go.wisc.edu/wj5665/)


---

class: middle

.center[
## Vertical Integration
]

---

### Motivation

1. Horizontal integration looks at the same features across many related datasets.

1. Vertical integration looks at the same dataset using many related assays.

.center[
&lt;img src="figures/vertical_integration.png" width=500/&gt;
]

---

### Example

.pull-left[
1. Connecting the dots:
  * Elevated inflammation is known to increased HIV risk.
  * Cervicovaginal bacteria are known to influence inflammation.
1. Earlier studies had only isolated individual bacteria (e.g., Gram-stained vaginal spears) and were not from representative, healthy populations.
]

.pull-right[
&lt;img src="figures/gosman-cover.png" width=500/&gt;
]

---

### Study Design: Data Generation

* **236** women age 18 - 23 from the FRESH cohort in Umlazi, South Africa. 
* **31** women acquired HIV during the 9 month course.

1. Microbiome Profiling
	1. 16S rRNA sequencing
	2. Virome (metagenomic sequencing)
2. Host Profiling
	1. Cytokines and chemokines (Multiplexed Bead Assay)
	2. Immune cell co-receptors (Flow Cytometry)
	3. Sexual behavior/vaginal hygiene survey

---

### CT and HIV Risk

They define 4 cervicotypes. 
  * _L. crispatus_ dominant (CT1, `\(n = 23\)`), _L. iners_ dominant (CT2, `\(n = 74\)`), _G. vaginalis_ dominant (CT3, `\(n = 68\)`), other dominant (CT4, `\(n = 70\)`).
  * None were related to host behavioral data.

&lt;img src="figures/gosman-cts.jpeg"/&gt;

---

### CT and HIV Risk

1. None of the CT1 women acquired HIV
2. CT4 was at higher risk. 95% interval was `\(HR \in \left[1.14, 17.27\right]\)`.

.center[
&lt;img src="figures/gosman-hr.jpeg" width=520/&gt;
]

---

### Immunological Mechanism

HIV targets CD4+ T cells with the CCR5 co-receptor. These cells were more common
in CT4.

.center[
&lt;img src="figures/gosman-corecepters.jpg" width=420/&gt;
]

---

### Immunological Mechanism

These "target" T cells are attracted by certain chemokines. These are elevated in CT4.

.center[
&lt;img src="figures/gosman-chemokines.jpg" width=400/&gt;
]

---

### Analysis Methods: Latent Variables

Latent variable methods seek shared structure `\(Z\)` across data modalities.

.pull-left[
&lt;img src="figures/latent_integration.png" width=550/&gt;
]

.pull-right[
Examples: Canonical Correlation Analysis, Multiple Factor Analysis, Partial Least Squares, Co-operative Learning

R Packages: `FactoMineR`, `mixOmics`, `multiview`.
]

---

### Analysis Methods: Network Inference

Network inference methods define edges within and between tables, showing how
measurements are related to one another.

.pull-left[
&lt;img src="figures/tree_integration.png" width=550/&gt;
]

.pull-right[
Examples: Random Intersection Trees, Iterative Random Forests.

R Packages: `FSInteract`, `iRF`.
]

---

### Challenges

1. Alignability: If one modality has little in common with the rest, then
including it forces us to ignore real shared variation in the other modalities.
We may be better off analyzing each dataset separately (Ma et al., 2024).

1. Differential Costs: Experimental design becomes especially complex when there
are multiple modalities and some are easier to collect than others.

---

### Simulator Definition

Today, we will re-estimate a copula covariance matrix across modalities,

``` r
merged_sim &lt;- sim1 |&gt;
  join_copula(sim2)
```

We could instead condition on a shared learned latent variable,

``` r
merged_sim &lt;- sim1 |&gt;
  join_pamona(sim2)
```
but this requires the `reticulate` package, which can be tricky to set up.

---

.center[
## Demo + Exercises
]

* Code repository: [go.wisc.edu/o5sn6w](https://go.wisc.edu/o5sn6w)
* Exercise solutions: [go.wisc.edu/yd3yz9](https://go.wisc.edu/yd3yz9)
* Live Link: [go.wisc.edu/wj5665](https://go.wisc.edu/wj5665/)

---

.center[
  ## Take Home Messages
]

---

### Why Simulate?

Simulators can be a sandbox for testing out analysis and research ideas.

1. **Power Analysis**: How many samples do we need? How can we best allocate our
samples across populations, timepoints, and technologies?

1. **Benchmarking**: Which methods work best under which conditions, and how
sensitive are they to changes in the simulation?

1. **Calibration**: Are we correctly interpreting the output of our workflows?

You can use simulation for more than proving your methods are the best...

---

### Building a Simulator

There are many simulation packages being released. They don't have to be black
boxes. Pay attention to:

  - Can it use a template dataset? What types?
  - How does it model individual features?
  - How does it model relationships beween features?
  - How is experimental or biological variation introduced?

When building a simulator, there will be trade-offs between faithfulness,
controllability, and generality of application. What matters for your use case?

---

### Evaluating a Simulator

We can visually check different aspects of simulator quality,

1. Featurewise visualization: Box plots, histograms, ECDFs.
1. Bivariate visualization: Pairs plots.
1. Multivariate visualization: PCA, heatmaps.

Use these to decide on new covariates or estimation routines, and note that they
can be translated into quantitative metrics (e.g., (Cao et al., 2021; Crowell et al., 2020)).

---

### Conclusion

1. Consider sharing your simulation code. There is a lot of effort that goes
into building simulators (for grant proposals, data analysis, and methods
development) that is unnecessarily duplicated across the community.

1. If you learn about or create a new type of simulator, let me know!

---


### References

Cao, Y. et al. (2021). "A benchmark study of simulation methods for single-cell RNA sequencing data". En. In: _Nat. Commun._ 12.1, p. 6911.

Crowell, H. L. et al. (2020). "Muscat detects subpopulation-specific state transitions from multi-sample multi-condition single-cell transcriptomics data". En. In: _Nat. Commun._
11.1, p. 6077.

Ma, R. et al. (2024). "Principled and interpretable alignability testing and integration of single-cell data". En. In: _Proc. Natl. Acad. Sci. U. S. A._ 121.10, p. e2313719121.

Nygaard, V. et al. (2016). "Methods that remove batch effects while retaining group differences may lead to exaggerated confidence in downstream analyses". En. In: _Biostatistics_
17.1, pp. 29-39.

---

Tonkin-Hill, G. (2023). _GitHub - gtonkinhill/TCGA\_analysis - github.com_. &lt;https://github.com/gtonkinhill/TCGA_analysis&gt;. [Accessed 21-06-2024].

Zhang, Z. et al. (2023). "Signal recovery in single cell batch integration". En. In: _bioRxiv.org_.
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
