---
title: ""
author: "Kris Sankaran"
output:
  xaringan::moon_reader:
    css: ["default", "css/xaringan-themer.css"]
    lib_dir: libs
    self_contained: false
    nature:
      highlightStyle: github
      highlightLines: true
      ratio: "16:9"
    seal: false
---

class: title

```{r, echo = FALSE, warning = FALSE, message = FALSE}
library(MASS)
library(knitr)
library(RefManageR)
library(tidyverse)
opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE, dpi = 200, fig.align = "center", fig.width = 6, fig.height = 3)
min_theme <- theme_minimal() + 
  theme(
    panel.grid.minor = element_blank(),
    panel.background = element_rect(fill = "#f7f7f7"),
    panel.border = element_rect(fill = NA, color = "#0c0c0c", size = 0.6),
    axis.text = element_text(size = 14),
    strip.text = element_text(size = 16),
    axis.title = element_text(size = 16),
    legend.position = "bottom"
  )
theme_set(min_theme)

# overwrite some default scales in ggplot2
scale_fill_continuous <- function(...) scico::scale_fill_scico(..., palette = "lapaz", direction = -1)
scale_colour_discrete <- function(...) ggplot2::scale_color_brewer(..., palette = "Set2")
scale_x_continuous <- function(...) ggplot2::scale_x_continuous(..., expand = c(0, 0))
scale_y_continuous <- function(...) ggplot2::scale_y_continuous(..., expand = c(0, 0))

BibOptions(
  check.entries = FALSE, 
  bib.style = "authoryear", 
  cite.style = "authoryear", 
  style = "markdown",
  hyperlink = FALSE, 
  dashed = FALSE,
  max.names = 1
)
bib <- ReadBib("references.bib")
```

<div id="title">
Semisynthetic Simulation for Biological Data Analysis
</div>
<div id="under_title">
Session 3: Integrative Analysis
</div>

<div id="subtitle">
Kris Sankaran <br/>
25 | June | 2024 <br/>
Melbourne Integrative Genomics<br/>
Lab: <a href="https://go.wisc.edu/pgb8nl">go.wisc.edu/pgb8nl</a> <br/>
</div>

<div id="subtitle_right">
<!-- Slides: <a href="https://go.wisc.edu/u10492">go.wisc.edu/u10492</a>  -->
</div>

---


### Session 3: Integrative Simulation

Learning Outcomes:

1. Explain and distinguish between the horizontal and vertical integration problems.
1. Discuss how simulation can improve validity and power in horizontal and vertical integration workflows.
1. Design simulators that respect batch or multi-omics structure to support
integration applications.

---

class: middle

.center[
## Horizontal Integration
]

---

### Problem

1. To maximize power, it helps to combine as many relevant datasets as possible.

1. Unfortunately, technical factors can cause systematic differences between
groups of samples. These are often called "batch effects."

.center[
  Horizontal Integration figure from NUS slides
]

---

### Sources of Batch Variation

These kind of unwanted variation can result from:

* Variation in experimental or sequencing protocols.

* Differences in library size.

* Variation in dataset storage methods.

* Uninteresting, but real, biological variation: situation in cell cycle,
differing proportions of dead cells.

.center[
]

---

### General Approaches

Many methods have been proposed to do this. They mostly fall into one of two
categories,

1. Correct: Normalize the data before any downstream analysis.
1. Account: Directly model batch-to-batch variation in the downstream analysis.

---

### General Approaches

We'll focus on correction methods, since they can benefit most directly from
simulation.

|--|---|---|
|  | Pros | Cons |
| Correct | Generality. Allows for any downstream analysis. | Can introduce its own artifacts. |
| Account | Cohesion. Only one method needed for everything.  | Time consuming. Analysis becomes "to big to fail." |

---

### Controls

A powerful idea in correction methods is to make proper use of controls:

* Replication: At specific stages of analysis, you can create replicates to clarify the nature of subsequent variation.
* Negative Controls: The experiment can be designed so that some features will not vary with interesting biological variation.

Effective replication and controls can help distinguish between biological
variation that happens upstream (in the original sample) and technical variation
that is introduced downstream (during the sample preparation and sequencing).

---
<!-- 
### Analogy: Agriculture

These ideas have classical analogs in agricultural field trials.

Replication $\iff$ Growing the same variety in separate fields
Negative Controls $\iff$ Growing a variety with no agricultural inputs

?
 -->

<!-- --- -->

---

### Challenges

Correction needs to be applied with care. It's easy for problems to arise.

.pull-left[
Overcorrection: Real biological variation can be masked by overly aggressive
batch effect correction methods.
]

.pull-right[

]

---

### Challenges

Correction needs to be applied with care. It's easy for problems to arise.

.pull-left[
Miscalibration: Plugging in batch-effect corrected outputs into differential
analysis workflows can lead to miscalibrated $p$-values. 
]

.pull-right[

]

---

### Simulation

Simulation can give insight into the behavior of correction methods and their
influence on downstream analysis. 

1. Overcorrection: We can simulate possible forms of biological and batch
variation and benchmark using this as ground truth.

1. Miscalibration: We can create data with known null and nonnull genes and
calculate power and false discovery rates.

---

### Exercise

---

class: middle

.center[
## Vertical Integration
]

---

### Problem

---

### References

```{r, results='asis', echo = FALSE}
PrintBibliography(bib, start = 1, end = 4)
```

---

```{r, results='asis', echo = FALSE}
PrintBibliography(bib, start = 5, end = 8)
```

---

```{r, results='asis', echo = FALSE}
PrintBibliography(bib, start = 9, end = 12)
```

---

```{r, results='asis', echo = FALSE}
PrintBibliography(bib, start = 13, end = 20)
```
