<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Intervention Analysis with Mirrors</title>
    <meta charset="utf-8" />
    <meta name="author" content="Kris Sankaran" />
    <script src="libs/header-attrs-2.27/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">


class: title

`\(\def\Dir{\text{Dir}}\)`
`\(\renewcommand{\exp}[1]{\operatorname{exp}\left(#1\right)}\)`
`\(\def\Mult{\text{Mult}}\)`
`\(\def\*#1{\mathbf{#1}}\)`
`\(\def\m#1{\boldsymbol{#1}}\)`
`\(\def\Unif{\text{Unif}}\)`
`\(\def\win{\tilde{w}_{\text{in}}}\)`
`\(\def\reals{\mathbb{R}}\)`
`\(\def\simplex{\Delta}\)`
`\(\def\*#1{\mathbf{#1}}\)`
`\(\def\m#1{\boldsymbol{#1}}\)`
`\(\def\PD{\mathrm{PD}}\)`
`\(\def\FDP{\mathrm{FDR}}\)`
`\(\newcommand\mbb[1]{\mathbb{#1}}\)`
`\(\newcommand\mbf[1]{\mathbf{#1}}\)`
`\(\def\mc#1{\mathcal{#1}}\)`
`\(\def\mrm#1{\mathrm{#1}}\)`
`\(\def\absarg#1{\left|#1\right|}\)`




## Microbiome Intervention Analysis with `mbtransfer`

&lt;div id="subtitle"&gt;
Kris Sankaran &lt;br/&gt;
06 | August | 2024 &lt;br/&gt;
Lab: &lt;a href="https://go.wisc.edu/pgb8nl"&gt;go.wisc.edu/pgb8nl&lt;/a&gt; &lt;br/&gt;
&lt;/div&gt;

&lt;div id="subtitle_right"&gt;
Methods for Microbiome Data Analysis&lt;br/&gt;
Joint Statistical Meetings&lt;/br&gt;
Slides: &lt;a href="https://go.wisc.edu/ta8m7b/"&gt;go.wisc.edu/ta8m7b&lt;/a&gt;&lt;br/&gt;
Paper: &lt;a href=" https://go.wisc.edu/5n5et7/"&gt;go.wisc.edu/5n5et7&lt;/a&gt;
&lt;/div&gt;

---

### Motivation: Today's Session

---

### Interaction

To control realistic microbiomes, we need ways of understanding how one change
can influence the rest of the community.


---

### Inference

We need test statistics to detect whether changes at different time lags are
meaningful. The key questions are:

1. **Who**: Which taxa are affected by a shift?
1. **When**: Are the effects immediate? Lagged? Do they persist?
1. **How**: Are there host factors that mediate the effect?

---

class: middle

.center[
## Approach
]

---

### Overview

1. Transfer Functions: Model the influence of environmental inputs and 
taxon-taxon interactions using a high-dimensional linear model. 

1. Mirror Statistics: Use agreement across data splits to allow false discovery
rate-controlled inference of perturbation effects.

---

### Transfer Functions

We wanted to further develop the intervention analysis techniques of Box et al. [1].

&lt;img src="figure/intervention-analysis.png" width="500" style="display: block; margin: auto;" /&gt;

---

### Transfer Functions

Classical transfer function models are a form of autoregressive model:

`$$y_{t} = \sum_{p = 1}^{P} A_{p} y_{t - p} + \sum_{q = 0}^{Q - 1} B_{q}w_{t - q} + \epsilon_{t}$$`

where `\(y_{t}\)` is the series of interest (e.g., ozone) and `\(w_{t}\)` encodes the intervention (e.g., new regulations).

---

### Transfer Functions

.pull-left[
* The pattern of autoregressive coefficients determines `\(y_{t}\)`'s response to
 "pulse" and "step" interventions.
* Note that, unlike generalized Lotka-Volterra models, they can model lagged
intervention effects.
]
.pull-right[
&lt;img src="figure/pulse.png" width="400" style="display: block; margin: auto;" /&gt;
]


---

### Transfer Functions for Microbial Communities

We estimate a separate transfer function model `\(f_{j}\)` for each taxon:

`$$\mathbf{y}_{t}^{(i)} = \mathbf{f}\left(\mathbf{Y}^{(i)}_{t - 1}, \mathbf{W}^{(i)}_{t}, \mathbf{z}^{(i)} \right) + \mathbf{\epsilon}_{t}^{(i)}$$`

For each coordinate, we use a regularized linear boosting model.

---

### Interpretation

* `\(\mathbf{y}_{t}^{(i)}\)`: The `\(i^{th}\)` series at time `\(t\)`.
* `\(\mathbf{Y}^{(i)}_{t - 1} \in \reals\)`: The previous `\(L\)` lags of the `\(i^{th}\)` series.
* `\(\mathbf{Z}^{(i)}_{t - 1} \in \reals\)`: An external series whose influence on the response on the microbiome community we would like to identify.

---

### Mirror Statistics

Which of the taxa `\(j\)` are effected by the external perturbation, and how long do
they last?


We will want test statistics that are sensitive to effects that are not easily
summarized by individual coefficients.

---

### General Approach: Mirror Statistics

The proposal of  first splits the data into `\(\mathcal{D}^{(1)} = \left(\mathbf{X}^{(1)}, \mathbf{y}^{(1)}\right)\)` and `\(\mathbf{D}^{(2)} = \left(\mathbf{X}^{(2)}, \mathbf{y}^{(2)}\right)\)` and check for agreement in the estimates `\(\hat{\mathbf{\beta}}^{(1)}\)` and `\(\hat{\mathbf{\beta}}^{(2)}\)` across splits.

&lt;img src="figure/mirror-explanation1.png" width=800/&gt;

---

### Test Statistic for Linear Model

To detect whether `\(\beta_{j} \neq 0\)`, we can use the statistic:

`\(M_{j} = \text{sign}\left(\hat{\beta}^{(1)}_{j}\hat{\beta}^{(2)}_{j}\right)\left[\left|\hat{\beta}_{j}^{(1)}\right| + \left|\hat{\beta}_{j}^{(2)}\right|\right]\)`

&lt;img src="figure/mirror-explanation1.png" width=800/&gt;

---

### FDR Estimation

To choose a significance threshold `\(t\)`, compute the empirical false discovery proportion:

`\(\widehat{\FDP}\left(t\right) = \frac{\left|\{j : M_{j} &gt; t\}\right|}{\left|\left\{j : M_{j} &lt; -t\right\}\right|}\)`

&lt;img src="figure/mirror-explanation2.png" width=800/&gt;

---

### FDR Estimation

&lt;img src="figure/mirror-explanation3.png" width=400/&gt;

---

### Partial Dependence


---

### Partial Dependence

For nonlinear effects, we first randomly split series into subsets, `\(\mathcal{D}^{(1)}\)` and `\(\mathcal{D}^{(2)}\)`. For each split `\(s\)`, we estimate models `\(\hat{\*f}^{(s)}\)` using `\(d_{t}^{(i)} = \left(\*y_{t + 1}^{(i)}, \*Y_{t}^{(i)}, \*W_{t + 1}^{(i)}, \*z^{(i)}\right)\)` for subjects `\(i\)` in that split and nonoverlapping segments beginning at times `\(t\)`. Next, we estimate the counterfactual difference between interventions `\(\tilde{\*W}_{t} = \*1_{Q}\)` and `\(\tilde{\*W}_{t} = \*0_{Q}\)` for each taxon:

`$$\PD_{j}^{(s)} = \frac{1}{\absarg{\mathcal{D}^{(s)}}}\sum_{d_{t}^{(i)} \in \mathcal{D}^{(s)}}\left[\hat{f}_{j}^{(s)}\left(\*Y^{(i)}_{t}, \*1_{Q}, \*z^{(i)}\right) - \hat{f}_{j}^{(s)}\left(\*Y^{(i)}_{t}, \*0_{Q}, \*z^{(i)}\right)\right].$$`

---

### Mirror Statistics

`\begin{align*}
	M_{j} = \text{sign}\left(\PD_{j}^{(1)}\PD_{j}^{(2)}\right)\left[\absarg{\PD_{j}^{(1)}} + \absarg{\PD_{j}^{(2)}}\right],
\end{align*}`

---

&lt;img src="figure/Fig1.png"/&gt;

---

## Experiments

---

### Questions

---

### Simulation Hyperparameters

* Phylogenetic correlation: Microbiome data often have correlated taxa due to shared evolutionary ancestry or occupation of similar ecological niches.

* Uneven sequencing depth: In a typical experiment, not all samples are sequenced to the same read depth. This necessitates sample-wise normalization.


---

### Generative Mechanism

here is some tex

`$$\mathbf{y}_{t}^{(i)}\vert\theta_{t}^{(i)},\mathbf{\varphi},b^{(i)}\sim\text{NB}\left(b^{(i)}\exp{\theta_{t}^{(i)}},\mathbf{\varphi}\right)\\\theta_{t}^{(i)}=\sum_{p=1}^{P}A_{p}\theta_{t-p}^{(i)}+\sum_{q=1}^{Q}\left(B_{q}+C_{q}\odot z^{(i)}\right)\mathbf{w}_{t-q}^{i}+\mathbf{\epsilon}_{t}^{(i)}\\b^{(i)}\sim\Gamma\left(10,\lambda\right)\\\epsilon_{t}^{(i)}\sim\mathcal{N}\left(0,\Sigma\right)$$`

and more text

---

### 

&lt;img src="figure/Fig6.png"/&gt;

---

&lt;img src="figure/simulation_results_config1.png" width=900/&gt;

---

&lt;img src="figure/simulation_results_config2.png" width=900/&gt;

---

&lt;img src="figure/simulation_results_config3.png" width=900/&gt;

---

&lt;img src="figure/Fig4.png" width=1100/&gt;


---


## Data Analysis

---


---

&lt;img src="figure/Fig7.png"/&gt;

---

&lt;img src="figure/Fig8.png"/&gt;

---

&lt;img src="figure/S1_Fig12.png"/&gt;


---

---

### Code Example


---

class: reference

### References

[1] G. E. Box et al. "Intervention analysis with applications to
economic and environmental problems". In: _Journal of the American
Statistical association_ 70.349 (1975), pp. 70-79.

---

class: reference

### References



---
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
