<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Lightning Talk 2021-09-08</title>
    <meta charset="utf-8" />
    <meta name="author" content="Kris Sankaran" />
    <script src="libs/header-attrs-2.8/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/ninjutsu.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">



.center[
# Latent Structure Laboratory
]

.pull-left[
&lt;img src="https://uwmadison.box.com/shared/static/96fo34yptxf9nqmu89xwsxbgknp1ps97.png" width=450px/&gt;
]

.pull-right[
All philosophy is based on two things only: curiosity and poor eyesight. The
trouble is, we want to know more than we can see.

-- Bernard de Fontanelle,
_Entretiens sur la pluralit√© des mondes_ (1686)
]

.center[
### UW Madison Statistics Lightning Talks
.large[Kris Sankaran | [krisrs1128.github.io/LSLab](krisrs1128.github.io/LSLab) | 08 September 2021]
]

---

### Data-Driven Methodology

.pull-left[
The Latent Structure Lab is motivated by scientific studies of the microbiome
and climate change

* **Microbiome: How should we describe the bacterial ecosystems that surround (and
inhabit) us?**

* Climate Change: What can we learn about the large scale Earth Systems impacted
by climate change?
]

.pull-right[
&lt;img src="https://uwmadison.box.com/shared/static/20mf8omlcj4dvs8nmh0lvqtdlvoz28rq.jpeg"/&gt;
]

---
### Data-Driven Methodology

.pull-left[
The Latent Structure Lab is motivated by scientific studies of the microbiome
and climate change

* Microbiome: How should we describe the bacterial ecosystems that surround (and
inhabit) us?

* **Climate Change: What can we learn about the large scale Earth Systems impacted
by climate change?**
]

.pull-right[
&lt;img src="https://uwmadison.box.com/shared/static/bl6kcoqntocdfn8n4389m6uol2pfedc0.png" width=360/&gt;
]

---
### Data-Driven Methodology

.pull-left[
The Latent Structure Lab is motivated by scientific studies of the microbiome
and climate change

* Microbiome: How should we describe the bacterial ecosystems that surround (and
inhabit) us?

* **Climate Change: What can we learn about the large scale Earth Systems impacted
by climate change?**
]

.pull-right[
&lt;img src="https://uwmadison.box.com/shared/static/i51kwdmgmdh18lbgbcuqso6vcknii4jq.png" width=360/&gt;
]

---
### Data-Driven Methodology

.pull-left[
The Latent Structure Lab is motivated by scientific studies of the microbiome
and climate change

* Microbiome: How should we describe the bacterial ecosystems that surround (and
inhabit) us?

* **Climate Change: What can we learn about the large scale Earth Systems impacted
by climate change?**
]

.pull-right[
&lt;img src="https://uwmadison.box.com/shared/static/dh8pss4xr9fy6grkxmmo83ww6pshmy5b.png" width=360/&gt;
]
---

### Data-Driven Methodology

.pull-left[
The Latent Structure Lab is motivated by scientific studies of the microbiome
and climate change

* Microbiome: How should we describe the bacterial ecosystems that surround (and
inhabit) us?

* **Climate Change: What can we learn about the large scale Earth Systems impacted
by climate change?**
]

.pull-right[
&lt;img src="https://uwmadison.box.com/shared/static/w9e5cgzmyg969e02pvuaoivpw852eu6z.png" width=360/&gt;
]
---


### Data Visualization

* We provide strategies for navigating complex, high-dimensional data
* We study ideas from dimensionality reduction and interactive visualization
* We aim to build tools that are easy to learn and use

.pull-left[
&lt;img src="https://uwmadison.box.com/shared/static/1wpeqd8prd5p7y6ls27sw4lk0w9feaas.png" width=550/&gt;
.caption[
Example interactive visualization for model comparison, by Xinran Miao.
]
]

.pull-right[
&lt;img src="https://uwmadison.box.com/shared/static/me0oi0lgwvc2qultgwsnuld53gurjo65.png" width=490/&gt;
.caption[
A few different types of glaciers, figure by Bruce Zheng.
]
]
---

### Latent Structure

Groupings and gradients can allow investigators to rapidly assess the overall
landscape of their data or model outputs.

.pull-left-slight[
&lt;img src="https://uwmadison.box.com/shared/static/1o50zi77zyfojpm0xvho4gewpahe4ufn.png"/&gt;
.caption[
Summarizing prototypes from a microbiome topic model.
]
]
.pull-right-slight[
&lt;img src="https://uwmadison.box.com/shared/static/ya6wf5gij6ejyde0jau9rkp5jo6te7d8.png"/&gt;
.caption[
Dynamic linking for spatial proteomics, by Tinghui Xu.
]
]

---

### Project 1: Multiscale Topic Visualization

* Biological systems are multiscale -- genes organize into pathways, cells
organize into tissues, bacteria organize into communities, ...
* We've studied meta-algorithms for visualizing ensembles of topic models
trained across several scales

.center[
&lt;img src="https://uwmadison.box.com/shared/static/9tndl7f06q543c165qodmhazqus2o7af.png" width=880/&gt;
]
---

### Project 1: Multiscale Topic Visualization

* You can try these ideas out in the R package `alto`
* [lasy.github.io/alto](lasy.github.io/alto)

&lt;img src="https://uwmadison.box.com/shared/static/r8g7mpfi36x3sexdhsp07wbkvxsng03g.png" width = 950/&gt;
---


### Project 2: Metabolomics in Model Microbiomes

.pull-left[
* Biological systems are multi-faceted, and combining several data modalities
gives a richer view than studying any one in isolation
* How does the microbiome influence the surrounding chemical environment?
* This is a key step towards engineering soil microbiomes
]

.pull-right[
&lt;img src="https://raw.githubusercontent.com/krisrs1128/talks/master/2021/20210820/figure/compound-2.png"/&gt;
]

---

### Project 2: Metabolomics in Model Microbiomes

.pull-left[
* Biological systems are multi-faceted, and combining several data modalities
gives a richer view than studying any one in isolation
* How does the microbiome influence the surrounding chemical environment?
* This is a key step towards engineering soil microbiomes
]

.pull-right[
&lt;img src="https://raw.githubusercontent.com/krisrs1128/talks/master/2021/20210820/figure/compound-1051.png"/&gt;
]

---

### Project 2: Metabolomics in Model Microbiomes

.pull-left[
* Biological systems are multi-faceted, and combining several data modalities
gives a richer view than studying any one in isolation
* How does the microbiome influence the surrounding chemical environment?
* This is a key step towards engineering soil microbiomes
]

.pull-right[
&lt;img src="https://raw.githubusercontent.com/krisrs1128/talks/master/2021/20210820/figure/compound-586.png"/&gt;
]

---
### Project 2: Metabolomics in Model Microbiomes

These are some of the estimated main effects from each fo 3882 metabolite-level
ANOVAs. Which of the metabolites are worth following up? We are studying how
some ideas from selective inference can be applied.

&lt;img src="https://github.com/krisrs1128/talks/blob/master/2021/20210820/figure/bf-scatter.png?raw=true"/&gt;

---

### Project 3: Monitoring Glacial Lakes

* We also apply data visualization principles to monitoring ecosystem change
* We have implemented algorithms for measuring change in glaciers and glacial
lakes
* To be useful to stakeholders, we need effective ways for navigating the
results

.center[
&lt;img src="https://uwmadison.box.com/shared/static/fhd8q1fixed7cc4m19g8me83bqfn3xvp.png" width=800/&gt;
]

---

### Project 3: Monitoring Glacial Lakes

* We also apply data visualization principles to monitoring ecosystem change
* We have implemented algorithms for measuring change in glaciers and glacial
lakes
* To be useful to stakeholders, we need effective ways for navigating the results

.center[
&lt;img src="https://uwmadison.box.com/shared/static/acfvua8hm0z8j187edd9mlsiocjyqacx.png" width=780/&gt;
.caption[
Lake time series, from an app by Sarah Tian.
]
]
---


### Reaching Out

You can learn more at [krisrs1128.github.io/LSLab](krisrs1128.github.io/LSLab).
I encourage you to reach out if you are interested in these topics, even if you
are still exploring.

Open office hours: 7225C (and zoom), Fridays 3-4pm

Email: ksankaran@wisc.edu

.center[
&lt;img src="https://uwmadison.box.com/shared/static/iu2ak50emn1bpwh9fytliv5ggqqq2dlz.png" width=340/&gt;
&lt;img src="https://uwmadison.box.com/shared/static/ebncvy25z4q6livvmzy9b10a1th5zt33.jpeg" width=250/&gt;
]
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
