\documentclass{beamer}
\usetheme{Warsaw}
\usepackage{natbib}
\usepackage{graphicx}
\input{preamble.tex}
\setbeamersize{text margin left=5pt,text margin right=5pt}

\setbeamerfont{institute}{size=\fontsize{7pt}{8pt}}
\setbeamerfont{date}{size=\fontsize{0pt}{0pt}}

\title{Methods for Regime Detection and Multitable Analysis}
\author{Kris Sankaran}

\begin{document}
\maketitle

\begin{frame}
  \frametitle{Problems of Interest}
 \begin{itemize}
 \item Regime Detection: Can we segment time intervals in a microbiome study in
   an unsupervised way?
   \begin{itemize}
   \item Intervals with similar dynamics should be placed in the same regime
   \item Can we characterize differences between taxa during these regimes?
   \end{itemize}
 \item Multitable Analysis: How should we simultaneously study multiple
   measurement types?
   \begin{itemize}
   \item How do different measurement types covary with one another?
   \item Compared to usual regression, when there are multiple related
     responses, there are fewer obvious practical approaches
   \end{itemize}
 \item Philosophy: Distill relevant themes, provide concrete workflows, and
   systematize data / study specific recommendations
 \end{itemize}
\end{frame}

\section{Regime Detection}
\label{sec:regime_detection}

\begin{frame}
  \frametitle{Recap}
 \begin{itemize}
 \item For this case study, we again use the antibiotics data of
   \citep{dethlefsen2008pervasive}
   \begin{itemize}
   \item 3 subjects at between 52 and 56 timepoints
   \item Two antibiotic time courses introduced around 12 - 16 and 40 - 44
   \item Long term effects in Subject F previously observed
   \end{itemize}
 \item Last time we approached the regime detection problem using hierarchical
   clustering with various distances, HMMs, Sticky HMMs, Infinite HMMs, and
   Switching Dynamical Systems
 \item We continue that survey using different approaches
 \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{CART}
  \begin{itemize}
  \item Order species according to a hierarchical
    clustering\footnote{Alternatively, use phylogenetic similarity.}, and
    consider the surface of species abundances across the time by species space
  \item CART can approximate this surface, providing a partition of timepoints
    and species in the process.
  \item Can be extended to binary data
  \item Limitation: Cannot update the species order (though you could imagine an
    iterative approach...)
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{CART}
\begin{itemize}
\item Structural model: The class of functions $\mathcal{F}$ that can be
  expressed as
\begin{align*}
  f\left(x\right) &= \sum_{m = 1}^{M} c_{m} \indic{x \in R_{m}},
\end{align*}
\item Fitting criteria: Minimize the expected generalization squared-error,
\begin{align*}
  \frac{1}{n} \sum_{i = 1}^{n} \left(y_{i} - c\left(x_{i}\right)\right)^{2},
\end{align*}
\item Estimation strategy: Recursively split rectangles into left ($l$) and
  right ($r$) subrectangles according to
\begin{align*}
  \left(j^{\ast}, t_{j1}^{\ast}\right) &= \arg \min_{\substack{j = 1, \dots, p \\ t_{j1} \in \reals}} \sum_{i \in R_{1, l}} \left(y_{i} - \bar{y}_{l}\right)^{2} + \sum_{i \in R_{1, r}} \left(y_{i} - \bar{y}_{r}\right)^{2}
\end{align*}
and prune to prevent overfitting
\end{itemize}
\end{frame}

\begin{frame}
  \frametitle{CART ($\asinh$ abundance)}
  \begin{figure}[ht]
    \centering
    \includegraphics[width=.8\textwidth]{figure/rpart_complex_2}
    \caption{Each panel represents a subject, the $x$-axis are individual
      species, the $y$-axis indexes time, and the color is the predicted
      abundance according to CART. Note the low abundance predictions during the
      antibiotic time course. \label{fig:rpart_complex} }
  \end{figure}
\end{frame}

\begin{frame}
  \frametitle{CART (Residual)}
  \begin{figure}[ht]
    \centering
    \includegraphics[width=.8\textwidth]{figure/rpart_complex_2_resid}
    \caption{An advantage of using a regression model is that we can directly
      inspect the quality of model fit, using
      residuals. \label{fig:rpart_complex_resid} }
  \end{figure}
\end{frame}

\begin{frame}
  \frametitle{Adjusting Model Complexity}
  \begin{figure}[ht]
    \centering
    \includegraphics[width=.8\textwidth]{figure/rpart_simple}
    \caption{By increasing the model complexity, we notice species with
      increasing abundances during the antibiotic time
      courses. \label{fig:rpart_simple} }
  \end{figure}
\end{frame}


\begin{frame}
  \frametitle{CART (Presence / Absence)}
  \begin{figure}[ht]
    \centering
    \includegraphics[width=.8\textwidth]{figure/rpart_binary}
    \caption{We can use the same method on presence-absence data, to identify
      whether certain regimes have many more exact zeros than others.
      \label{fig:rpart_binary} }
  \end{figure}
\end{frame}

\begin{frame}
  \frametitle{Aside: Sparse Time Series}
\begin{figure}[ht]
  \centering
  \includegraphics[width=0.9\textwidth]{figure/abt_zeros}
  \caption{One species from the antibiotics data set. It appears to have
    disappeared during the two antibiotics time courses. There is potentially
    value in modeling sequences of exact zeros in some regimes.
    \label{fig:abt_zeros} }
\end{figure}

\end{frame}

\begin{frame}
  \frametitle{Dynamic Tobit Model}
  \begin{itemize}
  \item To model sparsity in microbiome abundances, we can adopt a censoring
    point of view
  \item For example, the Dynamic Tobit model adds an extra layer of censoring on
    top of the ordinary state space model,
    \begin{align*}
      x_{t} &= Ax_{t - 1} + w_{t} \\
      \tilde{y}_{t} &= Cx_{t} + \eps_{t} \\
      y_{t} &= \tilde{y}_{t} \indic{\tilde{y}_{t} \geq c},
    \end{align*}
    where both $x$ and $\tilde{y}$ are unobserved.
  \item MCMC \citep{de1997scan, brockwell2006long, manrique1998simulation} and
    particle filtering \citep{andrieu2002particle} approaches are available
  \item Note this does not share information across species
  \end{itemize}
\end{frame}

\begin{figure}
  \frametitle{Antibiotics Example}
  \centering
  \includegraphics[width=\textwidth]{figure/abt_scan}
  \caption{Results of the scan sampler for a single species. Black points are
    observed data, red points are posterior samples, and the red line is a
    kernel smooth over samples. Implementation at
    \url{https://github.com/krisrs1128/readings/} \label{fig:abt_scan}
  }
\end{figure}

\begin{frame}
  \frametitle{Mixture of GPs}
 \begin{itemize}
 \item Gaussian Processes (GPs) provide an alternative to dynamical systems for
   working with functional data
 \item GPs are well-suited to homogeneous data, but cannot model regime
   switching directly, because the kernel bandwidth is constant
 \end{itemize}
\begin{figure}[ht]
  \centering
  \includegraphics[width=0.8\textwidth]{figure/gp_bandwidths}
  \caption{Different bandwidths in a GP are associated with different dynamics,
    but dynamics cannot switch across timepoints. \label{fig:gp_bandwidths} }
\end{figure}

\end{frame}

\begin{frame}
  \frametitle{Mixture of GPs}
\begin{itemize}
\item To adapt GPs to more heterogeneous situations, many mixture
  \citep{tresp2001mixtures, rasmussen2002infinite}, multiscale
  \citep{fox2012multiresolution, samostring}, and time-varying
  \citep{paciorek2003nonstationary} approaches have been proposed
\item We focus on the formulation of \ref{rasmussen2002infinite},
 \begin{align*}
   y\left(t_{i}\right) \vert \{ \text{class}\left(t_{i}\right) = k\} &\sim
   \Gsn\left(x_{k}\left(t_{i}\right), \sigma^{2}\right) \\
   x_{k}\left(t\right) &\sim \GP\left(0, \kappa_{k}\right)  \\
   \left[\text{class}\left(t_{i}\right)\right] &\sim \text{CRP}\left(\alpha\right)
 \end{align*}
\end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Aside: Implementation tricks}
  \begin{itemize}
  \item Inference is accomplished by alternating gibbs sampling of class
    assignments with HMC on kernel hyperparameters. Code at
    \url{https://github.com/krisrs1128/readings} (\texttt{igp\_mix} folder).
  \item Can check gibbs sampling conditionals by evaluating $\frac{p\left(x
    \vert z\right)}{p\left(x^{\prime} \vert z\right) &= \frac{p\left(x,
      z\right)}{p\left(x^{\prime}, z\right)}$ at two random points, since
    complete data likelihood is much easier to evaluate than conditional
    \citep{grosse2014testing}
  \item Can check HMC derivatives using numerical derivative calculations
  \item Can use generic HMC packages, plugging in problem specific densities /
    gradients
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Antibiotics Data}
\begin{figure}[ht]
  \centering
  \includegraphics[width=\textwidth]{figure/igp_samples}
  \caption{Each panel represents displays the posteriors at one
    iteration of the sampler. Posteriors for each process are displayed over the
    full data range, even when the associated timepoints occupy a relatively
    restricted range. \label{fig:igp_samples}}
\end{figure}

\begin{frame}
  \frametitle{Antibiotics Data}
\begin{figure}[ht]
  \centering
  \includegraphics[width=\textwidth]{figure/igp_classes}
  \caption{Each row is an iteration of the sampler, each column is a timepoint.
    Tiles are shaded by their class assignments. Note the label switching that
    occurs periodically. \label{fig:igp_classes} }
\end{figure}

\end{frame}

\begin{frame}
  \frametitle{Changepoint Detection}
  \begin{itemize}
  \item While there are many proposals for changepoint detection within
    individual series, few methods leverage information across multiple series
  \item \citep{fan2015empirical} proposes an approach (BASIC) for the
    multiple-series situation
  \item They have two insights that lead to statistical and computational
    performance improvements,
    \begin{itemize}
    \item Collapse: Perform sampling directly on cluster assignemnts
    \item Optimize: Can use a numerical Empirical Bayes procedure to avoid
      difficult prior elicitation
    \end{itemize}
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Changepoint Detection}
\begin{itemize}
\item Model changepoint probabilities $q_{t} &\sim \pi_{Q}$, but optimize $\pi_{Q}$ over a class
\begin{align*}
  \pi_{Q} &= \sum w_{k} \nu_{k}
\end{align*}
where $\nu_{k}$ are a prespecified dictionary of densities over $\left[0,
  1\right]$.
\item Alternately gibbs sample rows and columns, and perform (MCMC-corrected)
  swaps to improve mixing
\end{itemize}
\begin{figure}[ht]
  \centering
  \includegraphics[width=0.7\textwidth]{figure/basic_model}
  \caption{Figure 2 from \citep{fan2015empirical}, illustrating the assumed data
    generating process for the BASIC model. \label{fig:basic_model} }
\end{figure}
\end{frame}

\begin{frame}
  \frametitle{Antibiotics Data}
 \begin{figure}[ht]
   \centering
   \includegraphics[width=\textwidth]{figure/basic_heatmap}
   \caption{Each column is a species, shaded by its taxonomic family, rows are
     timepoints, and cells are shaded by their posterior changepoint probability
     $p\left(Z_{ij} = 1 \vert X\right)$. \label{fig:basic_heatmap} }
 \end{figure}
\end{frame}

\section{Multitable Analysis}
\label{sec:multitable analysis}

\begin{frame}
  \frametitle{WELL-China Study}
  \begin{itemize}
  \item Focused on the relationship between various indicators of wellness.
  \item Data about exercise, sleep, diet, mental health for 1969 people, in
    addition to 16s data on 221 participants (with plans for metabolomics,
    methylation, ...)
  \item We focus on question of body composition vs. microbiome structure
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Body Composition}
  \begin{itemize}
  \item Scans from DXA sensor to provide richer information than BMI alone
  \item Can we reproduce findings relating microbiome and BMI / metabolic
    syndrome, but using our richer data on body type?
    \begin{itemize}
    \item People with similar BMI might have different disease dispositions
      depending on the distribution of lean and fat mass
    \end{itemize}
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{PCA}

\end{frame}

\begin{frame}
  \frametitle{CCA}
\end{frame}

\begin{frame}
  \frametitle{Sparse CCA}
\end{frame}

\begin{frame}
  \frametitle{Probabilistic CCA}
\end{frame}

\begin{frame}
  \frametitle{CCA + LDA}
\end{frame}

\begin{frame}
  \frametitle{Conclusion}

\end{frame}
\bibliographystyle{plainnat}
\bibliography{refs.bib}

\section{Testing using Distances}
\label{sec:testing_distances}

\end{document}
