<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Department Lightning Talks 2024</title>
    <meta charset="utf-8" />
    <script src="libs/header-attrs-2.28/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">




&lt;div id="links"&gt;
Slides: https://go.wisc.edu/&lt;br&gt;
Lab Site: https://go.wisc.edu/pgb8nl
&lt;/div&gt;

### Microbiome Data Science

The goal of my lab is to help microbiome researchers get the most out of their
data. The essential statistical questions are:

* **Integration**: How should we analyze data gathered from multiple batches or
technologies?

* **Experimental Design**: How can we manipulate microbiomes so that they end up
in desirable community states?

* **Reproducibility**: How can we be sure our conclusions are trustworthy?

We work directly with microbiologists on problems related to HIV, the gut-brain
axis, and synthetic communities.

---

exclude: true

### Problem Solving

To solve these problems, we draw from classic ideas from statistics and computing:

* **Simulation**: It's easier to design experiments and benchmark methods when
we can quickly generate realistic data.

* **Visualization**: A good interface can shape the way we think for the better,
helping us be more more critical and creative.

---

### Themes: Simulation

We have written an online guide [1] for using simulators in microbiome analysis. We
used it to teach a short course for computational biologists.

&lt;img src="figures/microbiome_intro_chapter.png"/&gt;

---

### Themes: Simulation

.pull-three-quarters-left[
&lt;img src="figures/multivariate_power_curve.png" width=720/&gt;
]
.pull-three-quarters-right[
&lt;span style="font-size: 18px;"&gt;
Here is an example from the book about using a simulator to guide power analysis
in a multivariate model.  Panels A + B check simulator faithfulness, and panel C compares models across sample sizes.
&lt;/span&gt;
]

---

### Themes: Visualization

Topic models were independently developed for analyzing genotype and text
data [2; 3] and are now widely used in
computational genomics
[4; 5; 6; 7].

.center[
&lt;img src="figures/text_genotypes.png" width=1200/&gt;
]

---

### Themes: Visualization

1. They are helpful because real biological samples often don't separate cleanly into clusteres.

1. Instead of matching a sample to a single cluster centroid, view them as
mixtures of multiple representatives.

&lt;span style="font-size: 18px;"&gt;
.center[
&lt;img src="figures/clusters_vs_mixtures.png" width=565/&gt;&lt;br/&gt;
Figure adapted from [8].
]
&lt;/span&gt;

---

### Themes: Visualization

In [9], we studied how to navigate across an ensemble of topic models.

.center[
&lt;img src="figures/alto_sketches_annotated alignment.png" width="750" style="display: block; margin: auto;" /&gt;

&lt;span style="font-size: 20px;"&gt;
In the Sankey diagram, columns are models and rectangles are topics.
&lt;/span&gt;
]

---

### Themes: Visualization

The patterns of flow on this diagram are useful for deciding on the number of
topics present in a dataset.

.center[
&lt;img src="figures/lda-combined.png" width="2003" style="display: block; margin: auto;" /&gt;

&lt;span style="font-size: 20px;"&gt;
The diagnostics in this simulation suggest that the true `\(K\)` is 5.
&lt;/span&gt;
]

---

### Themes: Interface Design

We built an interface to specify and evaluate causal mediation analysis models.

.center[
&lt;img src="figures/multimedia_webpage.png"/&gt;
]
&lt;span style="font-size: 20px;"&gt;
The package is described in [10]. See also the package
and this blog post.
&lt;/span&gt;
]

---

### Themes: Interface Design

The interface makes it easy to share functions (e.g., sensitivity analysis)
across a range of model types.


``` r
library(multimedia)

model &lt;- multimedia(
    exper, # experiment data
    lnm_model(), # mediation model
    glmnet_model(lambda = 0.5) # outcome model
) |&gt;
    estimate(exper)
```

---

### Themes: Interface Design

This gives a principled approach to data integration. For example, these data
suggest disease `\(\to\)` microbe `\(\to\)` metabolite indirect effects.

.center[&lt;img src="figures/lasso_mediators_plot-1.png" width=740/&gt;]
.center[&lt;span style="font-size: 20px;"&gt; Each panel is a microbe-metabolite pair, and colors separate disease/healthy states.  &lt;/span&gt; ]

---

### Reaching Out

* You can learn more at [go.wisc.edu/pgb8nl](go.wisc.edu/pgb8nl).
    - Simulation: [11; 4; 1]
    - Interfaces: [10; 12]
    - Visualization: [9; 4; 13]

* I enjoy working with students with a variety of educational levels and backgrounds.

* Email: [ksankaran@wisc.edu](mailto:ksankaran@wisc.edu)

---
class: reference

### References

[1] D. M. Blei, A. Y. Ng, and M. I. Jordan. "Latent dirichlet allocation". In:
_J. Mach. Learn. Res._ 3.null (Mar. 2003), p. 993–1022. ISSN: 1532-4435.

[2] J. Fukuyama, K. Sankaran, and L. Symul. "Multiscale analysis of count data
through topic alignment". In: _Biostatistics_ 24.4 (Jun. 2022), p. 1045–1065.
ISSN: 1468-4357. DOI: 10.1093/biostatistics/kxac018.
&lt;http://dx.doi.org/10.1093/biostatistics/kxac018&gt;.

[3] A. Kim, S. Sevanto, E. R. Moore, et al. "Latent Dirichlet Allocation
modeling of environmental microbiomes". In: _PLOS Computational Biology_ 19.6
(Jun. 2023). Ed. by G. Zeller, p. e1011075. ISSN: 1553-7358. DOI:
10.1371/journal.pcbi.1011075.
&lt;http://dx.doi.org/10.1371/journal.pcbi.1011075&gt;.

[4] J. J. L. Kris Sankaran Saritha Kodikara and K. L. Cao. _Chapter 1
Introduction | Simulation for Microbiome Analysis - krisrs1128.github.io_.
&lt;https://krisrs1128.github.io/microbiome-simulation/index.html&gt;. [Accessed
09-09-2024].

[5] X. Peng, J. Lee, M. Adamow, et al. "A topic modeling approach reveals the
dynamic T cell composition of peripheral blood during cancer immunotherapy".
In: _Cell Reports Methods_ 3.8 (Aug. 2023), p. 100546. ISSN: 2667-2375. DOI:
10.1016/j.crmeth.2023.100546.
&lt;http://dx.doi.org/10.1016/j.crmeth.2023.100546&gt;.

[6] J. K. Pritchard, M. Stephens, and P. Donnelly. "Inference of Population
Structure Using Multilocus Genotype Data". In: _Genetics_ 155.2 (Jun. 2000),
p. 945–959. ISSN: 1943-2631. DOI: 10.1093/genetics/155.2.945.
&lt;http://dx.doi.org/10.1093/genetics/155.2.945&gt;.

[7] K. Sankaran. "Data Science Principles for Interpretable and Explainable
AI". In: _arXiv_ copyright = Creative Commons Attribution Non Commercial No
Derivatives 4.0 International (2024). DOI: 10.48550/ARXIV.2405.10552.
&lt;https://arxiv.org/abs/2405.10552&gt;.

[8] K. Sankaran and S. P. Holmes. "Generative Models: An Interdisciplinary
Perspective". In: _Annual Review of Statistics and Its Application_ 10.1 (Mar.
2023), p. 325–352. ISSN: 2326-831X. DOI:
10.1146/annurev-statistics-033121-110134.
&lt;http://dx.doi.org/10.1146/annurev-statistics-033121-110134&gt;.

[9] K. Sankaran and S. P. Holmes. "Latent variable modeling for the
microbiome". In: _Biostatistics_ 20.4 (Jun. 2018), p. 599–614. ISSN:
1468-4357. DOI: 10.1093/biostatistics/kxy018.
&lt;http://dx.doi.org/10.1093/biostatistics/kxy018&gt;.

[10] K. Sankaran and S. P. Holmes. "Multitable Methods for Microbiome Data
Integration". In: _Frontiers in Genetics_ 10 (Aug. 2019). ISSN: 1664-8021.
DOI: 10.3389/fgene.2019.00627. &lt;http://dx.doi.org/10.3389/fgene.2019.00627&gt;.

[11] L. Symul, P. Jeganathan, E. K. Costello, et al. "Sub-communities of the
vaginal microbiota in pregnant and non-pregnant women". In: _Proceedings of
the Royal Society B: Biological Sciences_ 290.2011 (Nov. 2023). ISSN:
1471-2954. DOI: 10.1098/rspb.2023.1461.
&lt;http://dx.doi.org/10.1098/rspb.2023.1461&gt;.

[12] C. Tataru, M. Peras, E. Rutherford, et al. "Topic modeling for multi-omic
integration in the human gut microbiome and implications for Autism". In:
_Scientific Reports_ 13.1 (Jul. 2023). ISSN: 2045-2322. DOI:
10.1038/s41598-023-38228-0. &lt;http://dx.doi.org/10.1038/s41598-023-38228-0&gt;.

---
class: reference

### References

[1] L. Symul, P. Jeganathan, E. K. Costello, et al. "Sub-communities of the
vaginal microbiota in pregnant and non-pregnant women". In: _Proceedings of
the Royal Society B: Biological Sciences_ 290.2011 (Nov. 2023). ISSN:
1471-2954. DOI: 10.1098/rspb.2023.1461.
&lt;http://dx.doi.org/10.1098/rspb.2023.1461&gt;.

[2] C. Tataru, M. Peras, E. Rutherford, et al. "Topic modeling for multi-omic
integration in the human gut microbiome and implications for Autism". In:
_Scientific Reports_ 13.1 (Jul. 2023). ISSN: 2045-2322. DOI:
10.1038/s41598-023-38228-0. &lt;http://dx.doi.org/10.1038/s41598-023-38228-0&gt;.

---

exclude: true

### Themes: Simulation

1. We have a more general review  of how simulation is used
across interdisciplinary research.

1. For example, this figure shows how a statistical emulator can be trained to
approximate an agent-based model of COVID-19 transmission.

.center[
&lt;img src="figures/emulation.jpeg"/&gt;
]

---

exclude: true

### Themes: Visualization

Recently, we have been studying how visualization can help interpret complex
machine learning models that are becoming adopted in microbiome research.

.center[
&lt;img src="figures/species_21_interpolation.svg" width=940/&gt;
]

---

exclude: true

### Microbiome Data Science

We work directly with biologists, and these questions have come up in our
projects together:

* **Integration**: Combine immunological (single-cell) and microbial
(metagenomic) data to understand factors that contribute to HIV risk.

* **Experimental Design**: Study how environmental factors influence whether an invading 
pathogen successfully attacks a synthetic root microbiome.

* **Reproducibility**: Gauge how antibiotics shape the microbiome at multiple
levels of taxonomic resolution.

---
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-0.15.0.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
