<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Visualizing Microbiome Data</title>
    <meta charset="utf-8" />
    <script src="libs/header-attrs-2.28/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">


# Visualizing Microbiome Data

&lt;div id="subtitle"&gt;
Kris Sankaran &lt;br/&gt;
AHRI Seminar &lt;br/&gt;
17 | January | 2025 &lt;br/&gt;
&lt;/div&gt;
&lt;div id="subtitle_right"&gt;
Slides: &lt;a href="https://go.wisc.edu/i50k4t"&gt;go.wisc.edu/i50k4t&lt;/a&gt;&lt;br/&gt;
Examples: &lt;a href="https://go.wisc.edu/9416ze"&gt;go.wisc.edu/9416ze&lt;/a&gt;&lt;br/&gt;
Lab: &lt;a href="https://go.wisc.edu/pgb8nl"&gt;go.wisc.edu/pgb8nl&lt;/a&gt; &lt;br/&gt;
&lt;/div&gt;

&lt;!-- 40 minute talk --&gt;


---

## Visualization Fundamentals

---

### Motivation

If our brains were built differently, we might be able to understand an entire
experiment by glancing at a spreadsheet. Data visualization allows us to use
our strengths in visual perception to our advantage.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/lion_binary.png" width=1200/&gt;
Figure from [1].
&lt;/div&gt;

---

### Motivation

If our brains were built differently, we might be able to understand an entire
experiment by glancing at a spreadsheet. Data visualization allows us to use
our strengths in visual perception to our advantage.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/paradoxes.png" width=1200/&gt;
Figure from [2].
&lt;/div&gt;

---

### Strategy

Statistical thinking allows us to get past this limitation:

1. **Compression**: Find a simple representation that captures essential
properties of the data. 

1. **Focusing**: Narrow down to a subset of data that is sufficient to
answer a question of interest. 

&lt;div style="display: flex; justify-content: center; max-width: 700px;margin: 0 auto"&gt;
    &lt;img src="figures/map.png" style="max-width: 70%; height: auto;"/&gt;
    &lt;img src="figures/microscope.png" style="max-width: 30%; height: auto;"/&gt;
&lt;/div&gt;

---

## Designing Effective Visualizations

---

### Encoding and Efficiency

Different ways of encoding information are perceived with different accuracies.
This means that any visualization implicitly prioritizes some comparisons over
others.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/ranking_encodings.png" width=600/&gt;&lt;br/&gt;
Figure from [3].
&lt;/div&gt;

---

How many times larger is the circle on the right?

&lt;div style="text-align: center;"&gt;
&lt;iframe src="figures/circles.html" style="height: 620px; width: 600px" frameborder="0"&gt;&lt;/iframe&gt;&lt;br/&gt;
&lt;/div&gt;

---

How many times taller is the bar on the right?

&lt;div style="text-align: center;"&gt;
&lt;iframe src="figures/bars.html" style="height: 620px" frameborder="0"&gt;&lt;/iframe&gt;&lt;br/&gt;
&lt;/div&gt;

---

### Data-to-Ink Ratio

&gt; Premature summarization is the root of all evil in statistics.
&gt;
&gt; -- Susan Holmes

.pull-left[
Good visualizations show more of the data, faithfully represent it, and are
memorable.
]

.pull-right[
&lt;div style="text-align: center;"&gt;
&lt;img src="figures/data_ink2.png" width=450/&gt;&lt;br/&gt;
Figure from [5].
&lt;/div&gt;
]

---

### Data-to-Ink Ratio

&gt; Premature summarization is the root of all evil in statistics.
&gt;
&gt; -- Susan Holmes

.pull-left[
Good visualizations show more of the data, faithfully represent it, and are
memorable.
]

.pull-right[
&lt;div style="text-align: center;"&gt;
&lt;img src="figures/data_ink1.png" width=500/&gt;&lt;br/&gt;
Figure from [5].
&lt;/div&gt;
]

---

exclude: true

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/memorable_visualizations.png" width=900/&gt;&lt;br/&gt;
Figure from [6].
&lt;/div&gt;

---

### Faceting

It's often surprisingly effective show many similar views in parallel. This is
the basis for techniques like small multiples and piling [5; 7]

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/biome_horizon.png" width=500/&gt;&lt;br/&gt;
Horizon plot made using the BiomeHorizon package [8].
&lt;/div&gt;

---

### Focus-plus-Context

We can let readers zoom into patterns of interest without losing relevant
context.

&gt; Overview first, zoom and filter, then details on demand.

-- Schneiderman's "Visual Information Seeking Mantra", from [9].

---

&lt;iframe src="https://krisrs1128.github.io/treelapse/pages/antibiotic.html#interpretation" style="height: 620px; width: 900px" frameborder="0"&gt;&lt;/iframe&gt;&lt;br/&gt;
Interactive sankey from the treelapse package [10].

---

## Microbiome Visualization

---

### MA Plot

An MA plot [11] is a good first step in a differential abundance analysis.

* Minus: Difference in log-abundances between groups.
* Average: Average of log-abundances across groups

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/ma_plot_da.png" width=700/&gt;
&lt;/div&gt;

---

### MA Plot

This is an MA plot comparing low vs. high BMI groups in the Atlas study of
healthy European adults [12].
&lt;div style="text-align: center;"&gt;
&lt;img src="figures/ma_plot_da.png" width=850/&gt;
&lt;/div&gt;

---

### ECDF Plots

If we want taxa-level detail, we can make an empirical cumulative distribution
function (ECDF) plot.  This traces `\(\left(t, \mathbf{P}\left(X \leq t\right)\right)\)` for across the range of observed values of `\(t\)`.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/ecdf_overall.png"/&gt;
&lt;/div&gt;

---

Unlike histograms, ECDFs don't require any choice of bin width. They 
perform surprisingly well in studies of uncertainty visualization [13].

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/histogram_narrow.png"/&gt;
&lt;/div&gt;

---

Unlike histograms, ECDFs don't require any choice of bin width. They 
perform surprisingly well in studies of uncertainty visualization [13].

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/histogram_wide.png"/&gt;
&lt;/div&gt;

---

### ECDF Interpretation

Here is an ECDF for the same BMI differential analysis from before. Curves further to the
right have systematically larger abundances.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/ecdf_testing.png"/&gt;
&lt;/div&gt;

---

### Considerations

1. Which taxa? We typically order the panels using the result of a differential
abundance analysis.

1. What scale? Different normalizations will prioritize some differences between
groups over others.

&lt;div style="text-align: center;"&gt;
    &lt;img src="figures/ecdf_overall.png"/&gt;
&lt;/div&gt;

---

### Considerations

1. Which taxa? We typically order the panels using the result of a differential
abundance analysis.

1. What scale? Different normalizations will prioritize some differences between
groups over others.

&lt;div style="text-align: center;"&gt;
    &lt;img src="figures/ecdf_log.png"/&gt;
&lt;/div&gt;

---

### Principal Components Analysis

Principal components analysis (PCA) gives an overview of all taxa
simultaneously. It searches for views that maximize variance.

&lt;div style="text-align: center;"&gt;
    &lt;img src="figures/pca_q1.png"/&gt;&lt;br/&gt;
    What is this object?
&lt;/div&gt;

---

### Principal Components Analysis

Principal components analysis (PCA) gives an overview of all taxa
simultaneously. It searches for views that maximize variance.

&lt;div style="text-align: center;"&gt;
    &lt;img src="figures/pca_q2.png"/&gt;&lt;br/&gt;
    Not so complicated now. Credit for this example goes to Julie Josse.
&lt;/div&gt;

---

### Principal Components Analysis

You can think of each sample as a high-dimensional vector across taxa. PCA finds
a low-dimensional view of those high-dimensional vectors.

&lt;div style="text-align: center;"&gt;
&lt;img src="https://github.com/krisrs1128/stat436_s24/blob/main/exercises/figure/pca_projection.png?raw=true" width=800/&gt;
&lt;/div&gt;

---

### Scree Plot

The principal components are sorted according to the amount of variance that
they explain. Pay attention to steep dropoffs -- these might suggest true
numbers of communities.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/pc_eigenvalues.png" width=800/&gt;
&lt;/div&gt;

---

### Resulting "Maps"

The samples are organized so that those with similar measurements are placed
close to one another. Different dimensions give different views.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/pc_scores_1-2.png" width=1200/&gt;
&lt;/div&gt;

---

### Resulting "Maps"

The samples are organized so that those with similar measurements are placed
close to one another. Different dimensions give different views.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/pc_scores_7-8.png" width=1200/&gt;
&lt;/div&gt;

---

### Components

The PCA components give a way of interpreting the axes of the map. The further
we move in the PC1 direction, the more we increase the positive PC1 taxa, and
vice versa for negative taxa.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/pca_loadings.png" width=1100/&gt;
&lt;/div&gt;

---
### PCA Interpretation

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/focus_pc_1.png"/&gt;
&lt;/div&gt;

---

### PCA Interpretation

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/focus_pc_2.png"/&gt;
&lt;/div&gt;

---

### Identifying Dimensions

Interesting variation can be buried in higher components. A ridgeplot [14] can help identify these.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/pc_ridge-4.png" width=1200/&gt;
&lt;/div&gt;

---

### Considerations



---

### Topic Models

&lt;div style="display: flex; justify-content: space-between; align-items: center;"&gt;
    &lt;div style="flex: 1; padding-right: 20px;"&gt;
        1. Since components can have both positive and negative values, effects in PCA
        can sometimes cancel out, complicating interpretations.&lt;br/&gt;&lt;br/&gt;
        1. Topic models deal with this by requiring components to be nonnegative
        [16; 17].
        Therefore, they can be interpreted as latent communities &lt;span
        class="cite" data-cite="sankaran2019latent"&gt;&lt;/span&gt;.
    &lt;/div&gt;
    &lt;img src="figures/latent_dirichlet_v2.png" width=600/&gt;
&lt;/div&gt;

---

### Topic Models

What were called components in PCA are now called _topics_. Each sample is a
mixture of topics.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/latent_dirichlet_v2.png" width=750/&gt;
&lt;/div&gt;

---

### STRUCTURE Plot

The mixtures across the entire dataset can be visualized in stacked barplot.
This dataset studied community dynamics after antibiotics treatment.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/antibiotic_memberships.png" width=2000/&gt;
&lt;/div&gt;

---

### Simulation-based Checks

Unlike PCA, topic models are a generative model. This allows us to evaluate model quality by comparing real with simulated samples. 

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/lda_posterior_predictive.jpeg" width=700/&gt;
&lt;/div&gt;

---

### Topics

We can also visualize the topics themselves. It helps to sort the taxa so that
those that have the highest variation across topics are shown first.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/antibiotic_superheat.png" width=1000/&gt;
&lt;/div&gt;

---

### Topics

We can also visualize the topics themselves. It helps to sort the taxa so that
those that have the highest variation across topics are shown first.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/beta_representatives.png" width=1000/&gt;
&lt;/div&gt;

---

### Alto Plot

The main hyperparameter in topic models is the number of topics `\(K\)`. It can be
helpful to compare the topic model results across a range of `\(K\)`
&lt;span class="cite" data-cite="Fukuyama2022"&gt;&lt;/span&gt;.

&lt;img src="figures/alto_sketches_annotated alignment.png"&gt;

---

### Alto Plot

Similar topics are often recovered across many `\(K\)`. Our package draws
larger edges between topics that have similar latent communities [18].

&lt;img src="figures/alto_sketches_annotated alignment.png"&gt;

---

### Alto Interpretation


&lt;div style="text-align: center;"&gt;
&lt;img src="figures/pregnancy_sankey.jpg" width=500/&gt;&lt;br/&gt;

Figure from [19].
&lt;/div&gt;

---

### MA Plot - Revisited

.pull-left[
1. MA plots are often helpful for identifying sources of technical variation,
like batch effects.

1. The example shown here compares microarray data from experiments that are
identical except for the choice of dye [20].
]

.pull-right[
&lt;div style="text-align: center;"&gt;
&lt;img src="figures/dye_effects.png" width=400/&gt;
&lt;/div&gt;
]

---

### CCA Biplots

Canonical Correlation Analysis (CCA) builds PCA-like maps for each dataset so
that they look as similar to one another as possible.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/sparse_cca_multitable.jpg" width=980/&gt;&lt;br/&gt;
Figure from [21].
&lt;/div&gt;

---

### CCA Components

Like in PCA, the directions in a CCA plot can be interpreted by analyzing the
CCA components. Each data source has its own set of CCA components.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/sparse_cca_multitable.jpg" width=980/&gt;&lt;br/&gt;
Figure from [21].
&lt;/div&gt;

---

### Considerations

1. Like with ECDF plots, the output from a dimensionality reduction will depend on
initial original data transformations.

1. When there are many more features than samples, PCA can be unstable. Regularized
versions, like the sparse PCA we used here [15], can be more stable [30; 31].


1. CCA searches for shared variation across tables. This can cause the results
to ignore table-specific variation.  To decompose variation into shared and
distinct components, consider methods like [22; 23].

---

### Mediation Analysis

If we have precise hypotheses about how datasets might be related, we can design
visualization to evaluate them.

For example, in mediation analysis, expect the treatment to affect an omic dataset indirectly through an intermediary.

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/mediation_diagram.jpg" width=500/&gt;
&lt;/div&gt;

---

### Mediation Analysis

This faceted plot shows the paths with the strongest indirect effects in joint
metabolome-microbiome analysis [24].

&lt;div style="text-align: center;"&gt;
&lt;img src="figures/mediation_scatterplot.jpg" width=900/&gt;
&lt;/div&gt;


---

### Conclusion

1. Visualization is helpful throughout the data analysis process --
validation, discovery, and communication.

1. The best visualizations are the product of repeated refinements, just like
the best writing.

1. Other helpful in microbiome research are residual plots [25], partial dependence [26; 27], phylogenetic [28],
and network visualizations [29].

---

class: reference

### References

[1] G. Alain et al. _Understanding intermediate layers using linear classifier probes_. 2017. URL:
[https://openreview.net/forum?id=ryF7rTqgl](https://openreview.net/forum?id=ryF7rTqgl).

[2] T. Munzner. _Perception, Frameworks, Color_. InfoVis MiniCourse, Lecture 1. Lecture. Presentation slides available at https://www.cs.ubc.ca/~tmm/talks/bord04.1/bord1.mgp.6up.pdf.
Jun. 2004.

[3] J. Heer et al. "Crowdsourcing graphical perception: using mechanical turk to assess visualization design". In: _Proceedings of the SIGCHI conference on human factors in computing
systems_. 2010, pp. 203-212.

[4] T. Munzner. _Visualization Analysis and Design_. AK Peters Visualization Series. London, England: CRC Press, Dec. 2014.

[5] E. R. Tufte et al. _The visual display of quantitative information_. Vol. 2. 9. Graphics press Cheshire, CT, 1983.

[6] M. A. Borkin et al. "What Makes a Visualization Memorable?" In: _IEEE Transactions on Visualization and Computer Graphics_ 19.12 (Dec. 2013), p. 2306–2315. ISSN: 1077-2626. DOI:
[10.1109/tvcg.2013.234](https://doi.org/10.1109%2Ftvcg.2013.234). URL: [http://dx.doi.org/10.1109/TVCG.2013.234](http://dx.doi.org/10.1109/TVCG.2013.234).

[7] F. Lekschas et al. "A generic framework and library for exploration of small multiples through interactive piling". In: _IEEE Transactions on Visualization and Computer Graphics_
27.2 (2020), pp. 358-368.

[8] I. Fink et al. "BiomeHorizon  : Visualizing Microbiome Time Series Data in R". In: _mSystems_ 7.3 (Jun. 2022). Ed. by M. G. I. Langille. ISSN: 2379-5077. DOI:
[10.1128/msystems.01380-21](https://doi.org/10.1128%2Fmsystems.01380-21). URL: [http://dx.doi.org/10.1128/msystems.01380-21](http://dx.doi.org/10.1128/msystems.01380-21).

[9] B. Shneiderman. "The eyes have it: A task by data type taxonomy for information visualizations". In: _The craft of information visualization_. Elsevier, 2003, pp. 364-371.

[10] K. Sankaran et al. "Interactive visualization of hierarchically structured data". In: _Journal of Computational and Graphical Statistics_ 27.3 (2018), pp. 553-563.

[11] S. Dudoit et al. "Statistical methods for identifying differentially expressed genes in replicated cDNA microarray experiments". In: _Statistica sinica_ (2002), pp. 111-139.

[12] L. Lahti et al. "Tipping elements in the human intestinal ecosystem". In: _Nature Communications_ 5.1 (Jul. 2014). ISSN: 2041-1723. DOI:
[10.1038/ncomms5344](https://doi.org/10.1038%2Fncomms5344). URL: [http://dx.doi.org/10.1038/ncomms5344](http://dx.doi.org/10.1038/ncomms5344).

[13] M. Kay et al. "When (ish) is My Bus?: User-centered Visualizations of Uncertainty in Everyday, Mobile Predictive Systems". In: _Proceedings of the 2016 CHI Conference on Human
Factors in Computing Systems_. CHI’16. ACM, May. 2016. DOI: [10.1145/2858036.2858558](https://doi.org/10.1145%2F2858036.2858558). URL:
[http://dx.doi.org/10.1145/2858036.2858558](http://dx.doi.org/10.1145/2858036.2858558).

---

class: reference

### References

[14] C. O. Wilke. "Package “ggridges”". In: _Ridgeline Plots in “ggplot2_ (2022).

[15] N. B. Erichson et al. _sparsepca: Sparse Principal Component Analysis (SPCA)_. R package version 0.1.2. 2018. URL:
[https://CRAN.R-project.org/package=sparsepca](https://CRAN.R-project.org/package=sparsepca).

[16] J. K. Pritchard et al. "Inference of population structure using multilocus genotype data". In: _Genetics_ 155.2 (2000), pp. 945-959.

[17] D. M. Blei et al. "Latent dirichlet allocation". In: _Journal of machine Learning research_ 3.Jan (2003), pp. 993-1022.

[18] J. Fukuyama et al. "Multiscale analysis of count data through topic alignment". In: _Biostatistics_ 24.4 (Jun. 2022), p. 1045–1065. ISSN: 1468-4357. DOI:
[10.1093/biostatistics/kxac018](https://doi.org/10.1093%2Fbiostatistics%2Fkxac018). URL:
[http://dx.doi.org/10.1093/biostatistics/kxac018](http://dx.doi.org/10.1093/biostatistics/kxac018).

[19] L. Symul et al. "Sub-communities of the vaginal microbiota in pregnant and non-pregnant women". In: _Proceedings of the Royal Society B_ 290.2011 (2023), p. 20231461.

[20] G. C. Tseng et al. "Issues in cDNA microarray analysis: quality filtering, channel normalization, models of variations and assessment of gene effects". In: _Nucleic acids
research_ 29.12 (2001), pp. 2549-2557.

[21] K. Sankaran et al. "Multitable methods for microbiome data integration". In: _Frontiers in genetics_ 10 (2019), p. 627.

[22] K. Z. Lin et al. "Quantifying common and distinct information in single-cell multimodal data with Tilted Canonical Correlation Analysis". In: _Proceedings of the National Academy
of Sciences_ 120.32 (2023), p. e2303647120.

[23] E. F. Palzer et al. "sJIVE: Supervised joint and individual variation explained". In: _Computational Statistics &amp;amp; Data Analysis_ 175 (Nov. 2022), p. 107547. ISSN: 0167-9473.
DOI: [10.1016/j.csda.2022.107547](https://doi.org/10.1016%2Fj.csda.2022.107547). URL: [http://dx.doi.org/10.1016/j.csda.2022.107547](http://dx.doi.org/10.1016/j.csda.2022.107547).

[24] H. Jiang et al. "Multimedia: multimodal mediation analysis of microbiome data". In: _Microbiology Spectrum_ (Dec. 2024). Ed. by J. Claesen. ISSN: 2165-0497. DOI:
[10.1128/spectrum.01131-24](https://doi.org/10.1128%2Fspectrum.01131-24). URL: [http://dx.doi.org/10.1128/spectrum.01131-24](http://dx.doi.org/10.1128/spectrum.01131-24).

---

### Figure Credits

Microscope by Hilmy Abiyyu Asad from &lt;a href="https://thenounproject.com/browse/icons/term/microscope/" target="_blank" title="Microscope Icons"&gt;Noun Project&lt;/a&gt; (CC BY 3.0)
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
