<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Enhancing Microbiome Analysis with Semisynthetic Data</title>
    <meta charset="utf-8" />
    <script src="libs/header-attrs-2.28/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">


class: title

# Enhancing Microbiome Analysis with Semisynthetic Data

&lt;style&gt;
.slide-background {
    background: url("figures/cover.png") no-repeat center center;
    background-size: cover;
    opacity: 0.5;
}
&lt;/style&gt;

&lt;div id="subtitle"&gt;
Kris Sankaran &lt;br/&gt;
Plant Pathology Seminar&lt;br/&gt;
18 | February | 2025 &lt;br/&gt;
&lt;/div&gt;
&lt;div id="subtitle_right"&gt;
Slides: &lt;a href="https://go.wisc.edu/689h7c"&gt;go.wisc.edu/689h7c&lt;/a&gt;&lt;br/&gt;
Lab: &lt;a href="https://go.wisc.edu/pgb8nl"&gt;go.wisc.edu/pgb8nl&lt;/a&gt; &lt;br/&gt;
&lt;/div&gt;

&lt;!-- 55 minute talk --&gt;



---

### Microbiome Data

1. A microbiome is a microbe-scale ecosystem.  It can be described by taxonomic
composition, genomic function, and biochemical features.

1. Advances in sequencing technology have made it easier than ever to rapidly
profile these taxonomic and genomic features in a range of sites, including in
the human body, on plant roots, and in the oceans. For this reason, they have
become an important part of medical, agricultural, and environmental questions.

---

### Statistical Challenges

Developing the data analysis for a microbiome study can be complicated by a
number of factors.

* **Integration**: How should we transform and analyze data across
batches or technologies, each with unique sources of technical variability?

* **Experimental Design**: How should we arrange sampling, assign treatments,
and place controls so that we can have powerful statistical conclusions?

* **Reproducibility**: How can we be sure our conclusions are trustworthy?

---

### Data Analysis Controversy 

.center[
    &lt;img src="figures/retraction.png"/&gt;
]

In June 2024, _Nature_ retracted a paper [1] the claimed
identify microbiome signatures of cancer. This came after one year's worth of
debate [2; 3]
about the data analysis.

---

### Data Analysis Controversy 

.center[
    &lt;img src="figures/retraction.png"/&gt;
]

The "disease signature" was an artifact resulting from the use of a batch effect
correction method.  Before we can understand the nuances of the story, we need
to learn about batch effects and correction methods.

---

### Simulation to Resolve Controversy

Gerry Tonkin-Hill has an excellent re-analysis [4] of the data from [1] which sheds light what were likely the source of the phantom
signals. The first part is a simulation.

.pull-left[
&lt;img src="figures/gth_sim1.png"/&gt;
]
.pull-right[
&lt;img src="figures/gth_sim2.png"/&gt;
]

---

### The Changing Simulation Landscape

Historically, microbiome researchers have only rarely used simulation in
their data analysis workflow.

  * Time Consuming: The simulators would have to be written from scratch, which requires significant effort.
  * Unrealistic: Even afterwards, the resulting data may not be realistic enough to use to guide any practical conclusions.

However, in the last few years, the situation has changed quite noticeably.

---

### Semisynthetic Data

One of the major advances has been the design of algorithms that can leverage public data resources.

* Semisynthetic Data: The output from a simulator that has been designed to mimic external, template data. 
* Template Data: Previously gathered experimental data that can be used to train a simulator.

It's no longer necessary to start from a blank slate.

---

### New Packages

We also have many more packages that implement these new methods. Here are 6 out
of the 11 packages discussed in our review [5].

.center[
  &lt;img src="figures/simulation_packages_table.png" width=850/&gt;
]
---

### Talk Outline

This talk gives examples of how semisynthetic data can help microbiome data
analysis. It is based on references 
[11; 12; 5].

&lt;hr/&gt;
&lt;div class="outline-container"&gt;
  &lt;div class="act-header"&gt;
    &lt;span class="tilde"&gt;&lt;/span&gt; Act I: Benchmarking and Power Analysis &lt;span class="tilde"&gt;&lt;/span&gt;
  &lt;/div&gt;
  &lt;div class="sub-item"&gt;Differential Abundance&lt;/div&gt;
  &lt;div class="sub-item"&gt;Dimensionality Reduction&lt;/div&gt;
  
  &lt;div class="interlude"&gt;
    &lt;span class="tilde"&gt;&lt;/span&gt; Interlude on Evaluation &lt;span class="tilde"&gt;&lt;/span&gt;
  &lt;/div&gt;
  
  &lt;div class="act-header"&gt;
    &lt;span class="tilde"&gt;&lt;/span&gt; Act II: Reliability and Attribution &lt;span class="tilde"&gt;&lt;/span&gt;
  &lt;/div&gt;
  &lt;div class="sub-item"&gt;Data Integration&lt;/div&gt;
  &lt;div class="sub-item"&gt;Mediation Analysis&lt;/div&gt;
&lt;/div&gt;
&lt;hr/&gt;

---

## Benchmarking and Power Analysis

---

### Differential Abundance

A common question in microbiome analysis is whether a given taxon is more vs.
less abundant in some conditions vs. others. Formally, consider
  * Hypotheses of interest: `\(H_{1}, \dots, H_{M}\)`. Some of them are non-null, but you don't know which.
  * Associated `\(p\)`-values: `\(p_{1}, \dots, p_{M}\)`.

Goal: Reject as many non-null hypotheses as possible while controlling the _False Discovery Rate_,

`\begin{align*}
\text{FDR} := \mathbf{E}\left[\frac{|\text{False Positives}|}{|\text{Rejections}| \vee 1}\right]
\end{align*}`

---

### Simulation Example

We can define benchmark using our data. In this example, we

1. Trained a simulator to mimic data from (tk: Give the study citation and dimensions)
1. Deliberately removed the effects from the (tk: how many) taxa with the
weakest effects according to a Wilcoxon signed rank test.

Step (2) can be thought of as a _computational negative control_. Together, we
can generate realistic data with ground truth signals.

---

### Example Simulated Data

The semisynthetic data seems to capture group and taxa differences among these
highly abundant taxa.

.center[
&lt;img src="figures/contrast-simulators-1.png" width=1000/&gt;
]

---

### Benchmarking Analysis

.pull-three-quarters-left[
&lt;img src="figures/visualize-da-power-1.png" width=700/&gt;
]

.pull-three-quarters-right[
All methods control the FDR. LIMMA has high variability in performance. Power
only plateaus around `\(n = 1000\)` samples.
]

---

### Implementation

For this simulator, we used a zero-inflated Negative Binomial variant of the
scDesign3 model [6]. For the
abundance of taxon `\(j\)` in sample `\(i\)`, we used:
`\begin{align*}
X_{ij} \sim \text{ZINB}\left(\mu_{g\left(i\right)j}, \varphi_{g\left(i\right)j}, \nu_{j}\right)
\end{align*}`
where `\(g\left(i\right)\)` is the BMI category of sample `\(i\)` and where 
`\(\mu, \varphi\)`, and `\(\nu\)` are mean, dispersion, and zero-inflation parameters,
respectively.

.center[
&lt;img src="figures/zinb_cartoon.png" width=350/&gt;
]

---

### Community-wide Associations

In many problems, we are interested in the relationships across a collection of
taxa. These analysis require more advanced methods, like network 
[7; 8] or dimensionality 
reduction [9] (see figure below) techniques.

.center[
&lt;img src="figures/antibiotic_prototypes.png" width=840/&gt;&lt;br/&gt;
&lt;span style="font-size: 18px;"&gt;
&lt;/span&gt;
]

---

### Motivation: Power Analysis

.pull-left[
1. Power analyses are intended to prevent researchers from embarking on studies
that have very little chance of detecting the hypothesized signals.

1. While there are formulas for certain univariate tests, there aren't any for 
more complex, multivariate models.
]

.pull-right[
&lt;img src="figures/lenth_power_calculator.png" width=500/&gt;
&lt;span style="font-size: 18px;"&gt;
Calculator from Russ Lenth's power and sample size webpage.
&lt;/span&gt;
]

---

### sPLS-DA Setting

Our power analysis uses Sparse Parital Least Squares Discriminant Analysis
(sPLS-DA) . This topic is it's
own full workshop, but let's review the core ideas.

.pull-left[
sPLS-DA helps with prediction when, 

* s: Not all features are predictive
* PLS: Many features are correlated with one another
* DA: The response is one of `\(K\)` classes
]

.pull-right[
&lt;img src="figures/PLS-setup.png" width=500/&gt;
]

---

### sPLS-DA Intuition

We "blend" columns of `\(\mathbf{X}\)` and `\(\mathbf{Y}\)` within tables until the patterns look similar.

.center[
&lt;img src="figures/PLS-step1.png" width=500/&gt;
]

Roughly, choose weights `\(\mathbf{a}\)` and `\(\mathbf{b}\)` to maximize
`\(\text{cor}\left(\mathbf{Xa}, \mathbf{Yb}\right)\)`.

---

### sPLS-DA Intuition

We "blend" columns of `\(\mathbf{X}\)` and `\(\mathbf{Y}\)` within tables until the patterns look similar.

.center[
&lt;img src="figures/PLS-step2.png" width=240/&gt;
]

Roughly, choose weights `\(\mathbf{a}\)` and `\(\mathbf{b}\)` to maximize
`\(\text{cor}\left(\mathbf{Xa}, \mathbf{Yb}\right)\)`.


---

### sPLS-DA Intuition

We "blend" columns of `\(\mathbf{X}\)` and `\(\mathbf{Y}\)` within tables until the patterns look similar.

.center[
&lt;img src="figures/PLS-step3.png" width=400/&gt;
]

Roughly, choose weights `\(\mathbf{a}\)` and `\(\mathbf{b}\)` to maximize
`\(\text{cor}\left(\mathbf{Xa}, \mathbf{Yb}\right)\)`.

---

### sPLS-DA Intuition

We "blend" columns of `\(\mathbf{X}\)` and `\(\mathbf{Y}\)` within tables until the patterns look similar.

.center[
&lt;img src="figures/PLS-step4.png" width=150/&gt;
]

Roughly, choose weights `\(\mathbf{a}\)` and `\(\mathbf{b}\)` to maximize
`\(\text{cor}\left(\mathbf{Xa}, \mathbf{Yb}\right)\)`.

---

### sPLS-DA Intuition

Now we can compare samples from the two tables in a single, shared space.

.center[
&lt;img src="figures/PLS-step5.png" width=800/&gt;
]

---

### sPLS-DA Intuition

Now we can compare samples from the two tables in a single, shared space.

.center[
&lt;img src="figures/PLS-step6.png" width=800/&gt;
]

---

### sPLS-DA Intuition

To get more than one dimension, we can repeat this process after removing any
correlation with previously found patterns.

.center[
&lt;img src="figures/PLS-step7.png" width=800/&gt;
]

---

### Example Output

In this example, we are comparing mice with and without a mouse model of Type I
diabetes (T1D). SPLS-DA helps us find taxa that distinguish healthy and disease
groups.

.pull-left[
&lt;img src="figures/t1d-true-data.png"/&gt;
]
.pull-right[
&lt;img src="figures/t1d-true-data-factors.png"/&gt;
]


---

### Problem Formulation

How many samples are necessary before this method can recover the
discriminating factors?

* **Estimate**: Train a simulator on the original data.
* **Alter/Sample**: Define negative control taxa with no association with T1D.
* **Gather/Summarize**: Evaluate SPLS-DA performance on semisynthetic data with
varying sample sizes and fractions of negative control taxa.

.center[
  &lt;img src="figures/simulation_steps.png" width=600/&gt;
]

---

### Bivariate Relationships

Except for the "streaks" when taxa were not observed, our simulator seems to
have preserved the essential bivariate relationships across taxa.
.center[
&lt;img src="figures/splsda-correlations.png" width=1000/&gt;
]

---

### Power Analysis

These are the results of our simulation experiment across varying sample sizes
and proportions of truly associated taxa. When few taxa are truly predictive,
many more samples are needed.

.center[
&lt;img src="figures/multivariate_power_isolated.png" width=700/&gt;
]

---

### Copula Models

These are a type of model that "couple" a collection of known marginal
distributions .

.center[
&lt;img src="figures/gene-gene_dependence.png" width=900/&gt;
]

---

### Starting Point

**Question**: If we were asked to simulate a vector of five correlated variables on
our computers right now, what would be the easiest thing to do?

---

### Starting Point

**Question**: If we were asked to simulate a vector of five correlated variables on
our computers right now, what would be the easiest thing to do?


```
##            [,1]       [,2]       [,3]      [,4]       [,5]
## [1,] -0.3889271 -0.2354183 -0.3853199 -0.488968 -0.3675065
## [2,] -1.7597348 -1.8336729 -2.0442459 -1.777946 -1.8570512
## [3,]  0.9878085  1.1354586  1.0483354  1.153748  1.0872393
```

The difficulty is that we usually want non-Gaussian margins `\(F_{1}, \dots, F_{D}\)`.

---

### Intuition

* In the Gaussianized space, it's easy to model correlation.
* The mapping back and forth is possible because we know the margins `\(F\)`.
  - `\(\Phi\)` represents the Gaussian CDF applied componentwise
&lt;br/&gt;
&lt;br/&gt;

.center[
&lt;img src="figures/copula_transformation.png" width=700/&gt;
]

---

### Copula Models

More formally, let `\(F_{1}, \dots, F_{D}\)` be the target margins and let `\(\Phi\)` be
the CDF of the Gaussian distribution. Gaussian Copula modeling has these steps.

Estimate:

1. Gaussianize the observed `\(\mathbf{x}_{i}\)` to `\(\mathbf{z}_{i} := \left[\Phi^{-1}\left(F_{1}\left(x_{i1}\right)\right), \dots, \Phi^{-1}\left(F_{D}\left(x_{iD}\right)\right)\right]\)`
1. Estimate the covariance `\(\hat{\Sigma}\)` associated with `\(z_{i}\)`

Simulate:

1. Draw `\(\mathbf{z}^\ast \sim \mathcal{N}\left(0, \Sigma\right)\)` 
1. Transform back `\(\mathbf{x}\ast := \left[F_{1}^{-1}\left(\Phi\left(z_{i1}^\ast\right)\right), \dots, F_{D}^{-1}\left(\Phi\left(z_{iD}^\ast\right)\right)\right]\)`

---

### Variations

1. We might expect the corelation structure to vary across groups. This can be
accomplished by setting separate `\(\Sigma_{k}\)` across groups `\(k\)`.

1. In high-dimensions, the sample covariance `\(\hat{\Sigma}\)` can destabilize. In
this case, we should use high-dimensional covariance estimators .

.center[
&lt;img src="figures/copula_groups.png" width=700/&gt;
]

---

## Interlude - Evaluation

---

### Evaluation Taxonomy

To be useful, simulated data need to be realistic. A few differences to be aware of:

* **Narrow/Broad Measures**: Narrow measures focus on small subsets of taxa, while broad measures evaluate community-level properties.

* **Graphical/Quantative**: Some checks are more easily quantifiable.

* **Fit-for-purpose measures**: Evaluation can focus on specific parameter estimates or analysis results.

Different types of realism should have higher priority depending on the
downstream tasks.

---

### Examples

* **Graphical, Narrow**: Boxplots or cumulative distribution function plots comparing real vs. simulated taxa.
* **Graphical, Broad**: Principal component plots of real vs. simulated dataset.
* **Quantitative, Narrow**: Two-sample Kolmogorov-Smirnov test.
* **Quantitative, Broad**: Performance of real vs. simulated data prediction.
* **Fit-for-Purpose**: Linear model coefficients on real vs. simulated data.

---

### Simulation Evaluation through Classification

Suppose we wanted to fit a model to these (simulated) data

&lt;center&gt;
&lt;img src="figures/true_mixture.png" width="500"/&gt;
&lt;/center&gt;

---

* A natural enough starting point is a Gaussian mixture model with `\(K = 4\)`.
* We can simulate from the fit, but it seems quite far off.
.pull-left[
_Simulated_
&lt;img src="figures/Gaussian (Shared Covarince).png" width="430"/&gt;
]
.pull-right[
_Truth_
&lt;img src="figures/true_mixture.png" width="430"/&gt;
]

---

We can make our assessment quantitative using the discriminator idea of .

The prediction probabilies below come from a GBM discriminator, which has an
out-of-sample prediction accuracy of 65.5%.

.pull-left[
_Simulated_
&lt;img src="figures/Gaussian (Shared Covarince)-prob.png" width="430"/&gt;
]

.pull-right[
_Truth_
&lt;img src="figures/true-Gaussian (Shared Covariance)-prob.png" width="430"/&gt;
]


---

As a next step,
* Increase number of components to `\(K = 5\)`.
* Fit different variances per component.

We still over-sample the gap between the two bottom-left clusters, but the GBM
accuracy has dropped down to 55.5%.
.pull-left[
_Simulated_&lt;br/&gt;
&lt;img src="figures/Gaussian (Individual Covariance)-prob.png" width="350"/&gt;
]
.pull-right[
_Truth_&lt;br/&gt;
&lt;img src="figures/true-Gaussian (Individual Covariance)-prob.png" width="350"/&gt;
]

---

* We use a mixture of `\(t\)` distributions next.
* GBM accuracy is now 50.6%
  - Unsurprisingly, this is the true mechanism that generated the data.

.pull-left[
_Simulated_

&lt;img src="figures/Student's t (Individual Covariance)-prob.png" width="350"/&gt;
]

.pull-right[
_Truth_

&lt;img src="figures/true-Student's t (Individual Covariance)-prob.png" width="350"/&gt;
]

---

The discrimination probabilities become closer to 0.5 the more accurate the simulation becomes.

&lt;center&gt;
&lt;img src="figures/modeling_iteration.png" width="800"/&gt;
&lt;/center&gt;

---

## Act 2: Reliability and Attribution

---

### Reliability Checks

1. Beyond power and benchmarking analysis, simulations can clarify how to
interpret a complicated workflow.

1. We have been calling this a *reliability check*.
These checks construct hypothetical scenarios to understand how methods behave.

&lt;div style="margin-left: 100px;"&gt;
&lt;span style="font-family: 'Exo 2'; font-size: 18;"&gt;
The analysis should not...&lt;br/&gt;
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;introduce spurious signals.&lt;br/&gt;
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;give high confidence results on uncertain data.&lt;br/&gt;
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;yield very different answers on similar datasets.&lt;br/&gt;
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;drown out subtle effects.&lt;br/&gt;
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;etc...
&lt;/span&gt;
&lt;/div&gt;

---

### Vertical Data Integration

To illustrate, let's consider a vertical data integration question. These are
problems where we get complementary 'omics views of the same samples.

.center[
&lt;img src="figures/vertical_integration.png" width="1000"/&gt;
]

The goal is to prepare a unified analysis, one which potentially considers
relationships across sources.

---

### ICU Example

.pull-left[
The study [10] used amplicon sequencing data to profile
the bacterial, viral, and fungal composition in gut microbiome samples from ICU
patients at a hospital, including a subset who were experiencing sepsis.
]

.pull-right[
&lt;img src="figures/transkingdom_summary.png" width="500"/&gt;
]

---

### Multiblock SPLS-DA Analysis

Multiblock SPLS-DA generalizes SPLS-DA to incorporate measurements across
multiple tables. With `\(\texttt{sepsis} \times \texttt{antibiotic use}\)` status as
the response variable, the method outputs the plots below.

.center[
&lt;img src="figures/multiblock_original.svg" width="1000"/&gt;
]

---

### Reliability Check

It's not obvious how we should interpret this output. The virus data must
influence the bacteria plot, because the method integrated the sources, but how
strong is the influence?

.center[
&lt;img src="figures/multiblock_original.svg" width="1000"/&gt;
]

Some integration methods are more vs. less aggressive than others.

---

### Semisynthetic Data

To calibrate our interpretation, we train a simulator using these data but then
deliberately remove any association between the bacteria community profile and
the sepsis response.

.center[
&lt;img src="figures/semisynthetic_transkingdom.png" width=820/&gt;
]
---

### Simulation Results

Applying Multiblock SPLS-DA to these data suggests that we are in an aggressive
integration regime. 
.center[&lt;img src="figures/multiblock_calibration.png" width=780/&gt;]
A reliability check like this might have helped [1]
realize that their normalization procedure introduced spurious associations.

---

## Mediation Analysis

---

### Model Comparison

It's common to compare models using `\(R^{2}\)` or prediction performance. Less well
known is that we can _also_ use semisynthetic data. This works even when
regression language is insufficient.

---

### Mindfulness Interventions

1. There is growing evidence for a relationship between the microbiome and
psychiatric conditions, both in mouse models and in observational human studies
.
2. Mindfulness interventions (e.g., meditation training) are known to alleviate
symptoms associated with depression and anxiety.
3. During one of its mindfulness training programs, the Center for Healthy Minds
gathered pre-post 16S data for both program participants and a waitlisted
control.

---

### Ongoing Collaboration

Jo Handelsman (Dept. Plant Pathology) and Richie Davidson (Depts. Psychology,
Psychiatry, and Neuroscience) have been leading several studies to explore these
connections in more depth.

.center[
&lt;img src="figures/crossroads.png" width=670/&gt;
]

---

### Pilot Study

We re-analyzed a pilot study from 2021, which gathered data from (tk) healthy
participants and gathered microbiome data before and after a mindfulness
intervention.

---

### Mediation Analysis

1. We were concerned that the mindfulness intervention might be affect behavior,
which in turn influences microbiota composition.
2. To explore this, we applied a form of mediation analysis to the 16S
microbiome and diet survey data .

.center[
&lt;img src="figures/mediation-dag.svg" width=450/&gt;
]

---

### Alternative Pathways

We can then simulate microbiome profiles assuming either:

1. Changes in the microbiome only result from diet changes.
1. Changes in the microbiome has no relationship with diet.

---

.center[
&lt;img src="figures/mindfulness-indirect-effects.png" width=900/&gt;
]

---

.center[
&lt;img src="figures/mindfulness-altered.png" width=800/&gt;
]

---


.center[
&lt;img src="figures/synthetic_null_testing.png" width=900/&gt;
]


---

## Conclusion

---

### Software and Tutorials

All the examples I discussed today can be run from online tutorials we've
written to accompany our papers:

* Simulation for Microbiome Analysis ([go.wisc.edu/wnj5p9](https://go.wisc.edu/wnj5p9))
* Generative Models Examples ([go.wisc.edu/ax73qb](https://go.wisc.edu/ax73qb))

The relevant R packages behind these analysis are:

* `multimedia` - Mediation analysis for microbiome data.
* `scDesign3` - An existing simulator for single cell data
* `scDesigner` - Under-development version used in the first tutorial.

---

Simulation can turn a problem of logic into one of observation.

.center[
&lt;img src="figures/simulation_summary.png" width=800/&gt;
]

We need your help! We are looking for more examples to help us write simulation
workflows for `scDesigner`. If you have data or problems that could benefit from
simulation, please reach out.

---

.center[
### Thank you!
]

* Contact: ksankaran@wisc.edu
* Lab Members: Margaret Thairu, Shuchen Yan, Yuliang Peng, Helena Huang
* Funding: NIGMS R01GM152744, NIAID R01AI184095

---

class: reference

### References

[1] G. D. Poore et al. "RETRACTED ARTICLE: Microbiome analyses of blood and tissues suggest cancer diagnostic approach". In: _Nature_ 579.7800 (Mar. 2020), p. 567–574.
ISSN: 1476-4687. DOI: [10.1038/s41586-020-2095-1](https://doi.org/10.1038%2Fs41586-020-2095-1). URL:
[http://dx.doi.org/10.1038/s41586-020-2095-1](http://dx.doi.org/10.1038/s41586-020-2095-1).

[2] A. Gihawi et al. "Major data analysis errors invalidate cancer microbiome findings". In: _mBio_ 14.5 (Oct. 2023). Ed. by I. B. Zhulin. ISSN: 2150-7511. DOI:
[10.1128/mbio.01607-23](https://doi.org/10.1128%2Fmbio.01607-23). URL: [http://dx.doi.org/10.1128/mbio.01607-23](http://dx.doi.org/10.1128/mbio.01607-23).

[3] G. D. Sepich-Poore et al. "Reply to: Caution Regarding the Specificities of Pan-Cancer Microbial Structure".  (Feb. 2023). DOI:
[10.1101/2023.02.10.528049](https://doi.org/10.1101%2F2023.02.10.528049). URL: [http://dx.doi.org/10.1101/2023.02.10.528049](http://dx.doi.org/10.1101/2023.02.10.528049).

[4] G. Tonkin-Hill. _GitHub - gtonkinhill/TCGA\_analysis - github.com_. &lt;https://github.com/gtonkinhill/TCGA_analysis&gt;. [Accessed 21-06-2024]. 2023.

[5] K. Sankaran et al. "Semisynthetic simulation for microbiome data analysis". En. In: _Brief. Bioinform._ 26.1 (Nov. 2024).

[6] D. Song et al. "scDesign3 generates realistic in silico data for multimodal single-cell and spatial omics". En. In: _Nature Biotechnology_ 42.2 (May. 2023), p.
247–252. DOI: [10.1038/s41587-023-01772-1](https://doi.org/10.1038%2Fs41587-023-01772-1). URL:
[http://dx.doi.org/10.1038/s41587-023-01772-1](http://dx.doi.org/10.1038/s41587-023-01772-1).

[7] Y. Shen et al. "Estimating sparse direct effects in multivariate regression with the spike-and-slab LASSO". In: _Bayesian Anal._ -1.-1 (Jan. 2024).

[8] Y. Shen et al. "The effect of the prior and the experimental design on the inference of the precision matrix in Gaussian chain graph models". En. In: _J. Agric. Biol.
Environ. Stat._ (May. 2024).

[9] K. Sankaran et al. "Latent variable modeling for the microbiome". In: _Biostatistics_ 20.4 (Jun. 2018), p. 599–614. ISSN: 1468-4357. DOI:
[10.1093/biostatistics/kxy018](https://doi.org/10.1093%2Fbiostatistics%2Fkxy018). URL:
[http://dx.doi.org/10.1093/biostatistics/kxy018](http://dx.doi.org/10.1093/biostatistics/kxy018).

[10] B. W. Haak et al. "Integrative transkingdom analysis of the gut microbiome in antibiotic perturbation and critical illness". En. In: _mSystems_ 6.2 (Mar. 2021).

[11] K. Sankaran et al. "Generative Models: An Interdisciplinary Perspective". In: _Annual Review of Statistics and Its Application_ 10.1 (Mar. 2023), p. 325–352. ISSN:
2326-831X. DOI: [10.1146/annurev-statistics-033121-110134](https://doi.org/10.1146%2Fannurev-statistics-033121-110134). URL:
[http://dx.doi.org/10.1146/annurev-statistics-033121-110134](http://dx.doi.org/10.1146/annurev-statistics-033121-110134).

[12] H. Jiang et al. "Multimedia: multimodal mediation analysis of microbiome data". In: _Microbiology Spectrum_ 13.2 (Feb. 2025). Ed. by J. Claesen. ISSN: 2165-0497.
DOI: [10.1128/spectrum.01131-24](https://doi.org/10.1128%2Fspectrum.01131-24). URL:
[http://dx.doi.org/10.1128/spectrum.01131-24](http://dx.doi.org/10.1128/spectrum.01131-24).

---

class: reference

### References



---

&lt;img src="figures/correlation_histogram.png" width=700/&gt;

---

&lt;img src="figures/covariance_hyperparameter_errors.png" width=700/&gt;
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
